<br/>

<img src="https://github.com/pachterlab/voyager/raw/documentation/vignettes/voyager.jpg" width="1024"/>

## Overview
This is the website for the spatial -omics Voyager workshop and hackathon at Caltech on March 4-8, 2024. Event info including schedule and workshop material can be found on this website.

### Description

Exploratory spatial data analysis (ESDA) can be a powerful approach to understanding single-cell genomics datasets, but it is not yet part of standard data analysis workflows. In particular, geospatial analyses, which have been developed and refined for decades, have yet to be fully adapted and applied to spatial single-cell analysis. We introduce the `Voyager` platform, which systematically brings the geospatial ESDA tradition to (spatial) -omics, with local, bivariate, and multivariate spatial methods not yet commonly applied to spatial -omics, united by a uniform user interface. Using `Voyager`, we showcase biological insights that can be derived with its methods, such as biologically relevant negative spatial autocorrelation. Underlying `Voyager` is the `SpatialFeatureExperiment` (SFE) data structure, which combines Simple Feature with `SingleCellExperiment` and `AnnData` to represent and operate on geometries bundled with gene expression data. Voyager has comprehensive tutorials demonstrating ESDA built on GitHub Actions to ensure reproducibility and scalability, using data from popular commercial technologies. Voyager is implemented in both R/Bioconductor and Python/PyPI, and features compatibility tests to ensure that both implementations return consistent results. 

In the workshop on March 4, you will get hands on experiences on the R implementation of SFE and Voyager. First, you will learn to create and operate on SFE objects. Next you will perform various types of ESDA on an SFE object with Voyager, and learn the spatial statistics behind these methods. The workshop material points you to further reading on these methods. You can read more about Voyager in [this preprint](https://www.biorxiv.org/content/10.1101/2023.07.20.549945v1?ct=). You will also learn to write R packages and use `git` version control.

In the hackathon on March 5-8, you will implement new features for SFE and Voyager as indicated in their GitHub repos and they are labeled with their difficulty. If you feel like all the issues are too challenging, you can bring your own spatial dataset for analysis. You may also open your own issues if you encounter bugs or want new features. Current Voyager crew (Lambda Moses and Alik Huseynov) will be here to answer questions. We will review your pull requests before late April for the Bioconductor 3.19 release. We plan to split up the Voyager preprint so there will be separate paper for SFE. Contributors will be added as coauthors of the relevant paper.
