<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="VoyagerHackathon">
<title>SpatialFeatureExperiment • VoyagerHackathon</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SpatialFeatureExperiment">
<meta property="og:description" content="VoyagerHackathon">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">VoyagerHackathon</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/info.html">Event info</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-workshop">Workshop</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-workshop">
    <a class="dropdown-item" href="../articles/sfe.html">SFE</a>
    <a class="dropdown-item" href="../articles/voyager.html">Voyager</a>
    <a class="dropdown-item" href="../articles/package.html">R package dev</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/VoyagerHackathon/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>SpatialFeatureExperiment</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:dlu2@caltech.edu" class="email">dlu2@caltech.edu</a>
      
                  
            <h4 data-toc-skip class="date">2024-03-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/VoyagerHackathon/blob/HEAD/vignettes/sfe.Rmd" class="external-link"><code>vignettes/sfe.Rmd</code></a></small>
      <div class="d-none name"><code>sfe.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="spatialfeatureexperiment">
<code>SpatialFeatureExperiment</code><a class="anchor" aria-label="anchor" href="#spatialfeatureexperiment"></a>
</h3>
<p><code>SpatialFeatureExperiment</code> (SFE) is a new <a href="http://adv-r.had.co.nz/S4.html" class="external-link">S4</a> class built on top of <a href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link"><code>SpatialExperiment</code></a>
(SPE). SFE incorporates geometries and geometric operations with the <a href="https://cran.r-project.org/web/packages/sf/index.html" class="external-link"><code>sf</code></a>
package. Examples of supported geometries are Visium spots represented
with polygons corresponding to their size, cell or nuclei segmentation
polygons, tissue boundary polygons, pathologist annotation of
histological regions, and transcript spots of genes. Using
<code>sf</code>, <code>SpatialFeatureExperiment</code> leverages the
GEOS C++ library underlying <code>sf</code> for geometry operations,
including algorithms for for determining whether geometries intersect,
finding intersection geometries, buffering geometries with margins, etc.
A schematic of the SFE object is shown below:</p>
<p>Below is a list of SFE features that extend the SPE object:</p>
<ul>
<li>
<code>colGeometries</code> are <code>sf</code> data frames
associated with the entities that correspond to columns of the gene
count matrix, such as Visium spots or cells. The geometries in the
<code>sf</code> data frames can be Visium spot centroids, Visium spot
polygons, or for datasets with single cell resolution, cell or nuclei
segmentations. Multiple <code>colGeometries</code> can be stored in the
same SFE object, such as one for cell segmentation and another for
nuclei segmentation. There can be non-spatial, attribute columns in a
<code>colGeometry</code> rather than <code>colData</code>, because the
<code>sf</code> class allows users to specify how attributes relate to
geometries, such as “constant”, “aggregate”, and “identity”. See the
<code>agr</code> argument of the <a href="https://r-spatial.github.io/sf/reference/sf.html" class="external-link"><code>st_sf</code>
documentation</a>.</li>
<li>
<code>colGraphs</code> are spatial neighborhood graphs of cells or
spots. The graphs have class <code>listw</code> (<code>spdep</code>
package), and the <code>colPairs</code> of
<code>SingleCellExperiment</code> was not used so no conversion is
necessary to use the numerous spatial dependency functions from
<code>spdep</code>, such as those for Moran’s I, Geary’s C, Getis-Ord
Gi*, LOSH, etc. Conversion is also not needed for other classical
spatial statistics packages such as <code>spatialreg</code> and
<code>adespatial</code>.</li>
<li>
<code>rowGeometries</code> are similar to
<code>colGeometries</code>, but support entities that correspond to rows
of the gene count matrix, such as genes. A potential use case is to
store transcript spots for each gene in smFISH or in situ sequencing
based datasets.</li>
<li>
<code>rowGraphs</code> are similar to <code>colGraphs</code>. A
potential use case may be spatial colocalization of transcripts of
different genes.</li>
<li>
<code>annotGeometries</code> are <code>sf</code> data frames
associated with the dataset but not directly with the gene count matrix,
such as tissue boundaries, histological regions, cell or nuclei
segmentation in Visium datasets, and etc. These geometries are stored in
this object to facilitate plotting and using <code>sf</code> for
operations such as to find the number of nuclei in each Visium spot and
which histological regions each Visium spot intersects. Unlike
<code>colGeometries</code> and <code>rowGeometries</code>, the number of
rows in the <code>sf</code> data frames in <code>annotGeometries</code>
is not constrained by the dimension of the gene count matrix and can be
arbitrary.</li>
<li>
<code>annotGraphs</code> are similar to <code>colGraphs</code> and
<code>rowGraphs</code>, but are for entities not directly associated
with the gene count matrix, such as spatial neighborhood graphs for
nuclei in Visium datasets, or other objects like myofibers. These graphs
are relevant to <code>spdep</code> analyses of attributes of these
geometries such as spatial autocorrelation in morphological metrics of
myofibers and nuclei. With geometry operations with <code>sf</code>,
these attributes and results of analyses of these attributes
(e.g. spatial regions defined by the attributes) may be related back to
gene expression.</li>
<li>
<code>localResults</code> are similar to <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations" class="external-link"><code>reducedDims</code>
in <code>SingleCellExperiment</code></a>, but stores results from
univariate and bivariate local spatial analysis results, such as from <a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link"><code>localmoran</code></a>,
<a href="https://r-spatial.github.io/spdep/reference/localG.html" class="external-link">Getis-Ord
Gi*</a>, and <a href="https://r-spatial.github.io/spdep/reference/LOSH.html" class="external-link">local
spatial heteroscedasticity (LOSH)</a>. Unlike in
<code>reducedDims</code>, for each type of results (type is the type of
analysis such as Getis-Ord Gi*), each feature (e.g. gene) or pair of
features for which the analysis is performed has its own results. The
local spatial analyses can also be performed for attributes of
<code>colGeometries</code> and <code>annotGeometries</code> in addition
to gene expression and <code>colData</code>. Results of multivariate
spatial analysis such as <a href="https://cran.r-project.org/web/packages/adespatial/vignettes/tutorial.html#multispati-analysis" class="external-link">MULTISPATI
PCA</a> can be stored in <code>reducedDims</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="voyager">
<code>Voyager</code><a class="anchor" aria-label="anchor" href="#voyager"></a>
</h3>
</div>
<div class="section level3">
<h3 id="x-visium">10X Visium<a class="anchor" aria-label="anchor" href="#x-visium"></a>
</h3>
<p>As Visium from 10X Genomics is currently the most popular spatial
transcriptomics technology, this workshop uses a Visium dataset.</p>
<p>In Visium, capture sequences with spot barcode, unique molecule
identifier (UMI), and polyT to capture poly-adenylated mRNAs are printed
in a hexagonal array on a glass slide. Each spot barcode has a known
location, and the spots are 55 <span class="math inline">\(\mu
m\)</span> in diameter and 100 <span class="math inline">\(\mu
m\)</span> apart center to center. As the spots are much larger than
most types of cells, Visium does not have single cell resolution. Tissue
is mounted on each of the 4 capture areas on the slide, and each capture
area has 4992 spots. The spots capture the transcripts from the tissue,
which are then reverse transcribed, amplified, and sequenced.</p>
<p>Space Ranger is the official software to align the sequencing reads
to the genome and quantify the UMIs in each spot for each gene. Spatial
Ranger also takes in a histology image of the capture area, with which
it determines which spots are in tissue. The user can also manually
determine which spots are in tissue in the Loupe Browser.</p>
</div>
<div class="section level2">
<h2 id="part-1-the-sfe-class">Part 1: The SFE class<a class="anchor" aria-label="anchor" href="#part-1-the-sfe-class"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="create-an-sfe-object">Create an SFE object<a class="anchor" aria-label="anchor" href="#create-an-sfe-object"></a>
</h3>
<p>10x Genomics Space Ranger output from a Visium experiment can be read
in a similar manner as in <code>SpatialExperiment</code>; the
<code>SpatialFeatureExperiment</code> SFE object has the
<code>spotPoly</code> column geometry for the spot polygons. If the
filtered matrix (i.e. only spots in the tissue) is read in, then a
column graph called <code>visium</code> will also be present for the
spatial neighborhood graph of the Visium spots on the tissue. The graph
is not computed if all spots are read in regardless of whether they are
on tissue.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample01"</span>, <span class="st">"sample02"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The results for each tissue capture should be in the
<code>outs</code> directory under the sample directory. Inside the
<code>outs</code> directory, these directories may be present:
<code>raw_reature_bc_matrix</code> has the unfiltered gene count matrix,
<code>filtered_feature_bc_matrix</code> has the gene count matrix for
spots in tissue, and <code>spatial</code> has the spatial information.
The matrix directories contain the matrices in MTX format as sparse
matrices. Space Ranger also outputs the matrices as h5 files, which are
read into R in a similar way as MTX.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Inside the matrix directory:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span>, <span class="st">"filtered_feature_bc_matrix"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Inside the <code>spatial</code> directory:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><code>tissue_lowres_image.png</code> is a low resolution image of the
tissue. Not all Visium datasets have all the files here. The
<code>barcode_fluorescence_intensity.csv</code> file is only present in
datasets with fluorescent imaging rather than bright field H&amp;E.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, sample_id <span class="op">=</span> <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, </span>
<span>                          data <span class="op">=</span> <span class="st">"filtered"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Space Ranger output includes the gene count matrix, spot coordinates,
and spot diameter. The Space Ranger output does NOT include nuclei
segmentation or pathologist annotation of histological regions. Extra
image processing, such as with ImageJ and QuPath, are required for those
geometries.</p>
<p>See <a href="https://pachterlab.github.io/voyager/articles/create_sfe.html" class="external-link">this
vignette</a> on creating SFE objects from scratch and for other spatial
trancriptomics technologies.</p>
</div>
<div class="section level3">
<h3 id="operations-of-sfe-objects">Operations of SFE objects<a class="anchor" aria-label="anchor" href="#operations-of-sfe-objects"></a>
</h3>
<p>User interfaces to get or set the geometries and spatial graphs
emulate those of <code>reducedDims</code> and <code>row/colPairs</code>
in <code>SingleCellExperiment</code>. Column and row geometries also
emulate <code>reducedDims</code> in internal implementation, while
annotation geometries and spatial graphs differ. Operations on SFE
objects are demonstrated on a small toy dataset (you may need to answer
a prompt in the R console when downloading the dataset):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="column-geometries">Column geometries<a class="anchor" aria-label="anchor" href="#column-geometries"></a>
</h4>
<p>Column geometries or <code>colGeometries</code> are the geometries
that correspond to columns of the gene count matrix, such as Visium
spots and cells in datasets from a single cell resolution technology.
Each SFE object can have multiple column geometries. For example, in a
dataset with single cell resolution, whole cell segmentation and nuclei
segmentation are two different <code>colGeometries</code>. However, for
Visium, the spot polygons are the only <code>colGeometry</code>
obviously relevant, though users can add other geometries such as
results of geometric operations on the spot polygons. The different
geometries can be get or set with their names, and “spotPoly” is the
standard name for Visium spot polygons.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get Visium spot polygons</span></span>
<span><span class="op">(</span><span class="va">spots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"spotPoly"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">st_geometry</span><span class="op">(</span><span class="va">spots</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set colGeometry</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots</span></span></code></pre></div>
<p>To see which <code>colGeometries</code> are present in the SFE
object:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">colGeometryNames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p>There are shorthands for some specific column or row geometries. For
example, <code>spotPoly(sfe)</code> is equivalent to
<code>colGeometry(sfe, "spotPoly")</code> shown above.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">spots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h4>
<p>Annotation geometries can be get or set with
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry()</a></code>. In column or row geometries, the number of
rows of the <code>sf</code> data frame (i.e. the number of geometries in
the data frame) is constrained by the number of rows or columns of the
gene count matrix respectively, because just like <code>rowData</code>
and <code>colData</code>, each row of a <code>rowGeometry</code> or
<code>colGeometry</code> <code>sf</code> data frame must correspond to a
row or column of the gene count matrix respectively. In contrast, an
<code>annotGeometry</code> <code>sf</code> data frame can have any
dimension, not constrained by the dimension of the gene count
matrix.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter, by name or index</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"tissueBoundary"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">st_geometry</span><span class="op">(</span><span class="va">tb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter, by name or index</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span></span></code></pre></div>
<p>See which <code>annotGeometries</code> are present in the SFE
object:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p>There are shorthands for specific annotation geometries. For example,
<code>tissueBoundary(sfe)</code> is equivalent to
<code>annotGeometry(sfe, "tissueBoundary")</code>.
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">cellSeg()</a></code> (cell segmentation) and <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">nucSeg()</a></code>
(nuclei segmentation) would first query <code>colGeometries</code> (for
single cell, single molecule technologies, equivalent to
<code>colGeometry(sfe, "cellSeg")</code> or
<code>colGeometry(sfe, "nucSeg")</code>), and if not found, they will
query <code>annotGeometries</code> (for array capture and
microdissection technologies, equivalent to
<code>annotGeometry(sfe, "cellSeg")</code> or
<code>annotGeometry(sfe, "nucSeg")</code>).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="spatial-graphs">Spatial graphs<a class="anchor" aria-label="anchor" href="#spatial-graphs"></a>
</h4>
<p>The spatial neighborhood graphs for Visium spots are stored in the
<code>colGraphs</code> field, which has similar user interface as
<code>colGeometries</code>. SFE also wraps all methods to find the
spatial neighborhood graph implemented in the <code>spdep</code>
package, and triangulation is used here as demonstration.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">"tri2nb"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">g</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph by name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"graph1"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">g</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get graph by name</span></span>
<span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"graph1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For Visium, spatial neighborhood graph of the hexagonal grid can be
found with the known locations of the barcodes. One SFE object can have
multiple <code>colGraphs</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Which graphs are present in this SFE object?</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraphNames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p>While this workshop only works with one sample, i.e. tissue section,
operations on multiple samples is discussed in <a href="https://pachterlab.github.io/SpatialFeatureExperiment/articles/SFE.html#multiple-samples" class="external-link">the
vignette of the SFE package</a>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Alik Huseynov, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
