<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="VoyagerHackathon">
<title>SpatialFeatureExperiment • VoyagerHackathon</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SpatialFeatureExperiment">
<meta property="og:description" content="VoyagerHackathon">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">VoyagerHackathon</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/info.html">Event info</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-workshop">Workshop</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-workshop">
    <a class="dropdown-item" href="../articles/sfe.html">SFE</a>
    <a class="dropdown-item" href="../articles/voyager.html">Voyager</a>
    <a class="dropdown-item" href="../articles/package.html">R package dev</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/lambdamoses/VoyagerHackathon/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>SpatialFeatureExperiment</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:dlu2@caltech.edu" class="email">dlu2@caltech.edu</a>
      
                  
            <h4 data-toc-skip class="date">2024-03-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lambdamoses/VoyagerHackathon/blob/HEAD/vignettes/sfe.Rmd" class="external-link"><code>vignettes/sfe.Rmd</code></a></small>
      <div class="d-none name"><code>sfe.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="spatialfeatureexperiment">
<code>SpatialFeatureExperiment</code><a class="anchor" aria-label="anchor" href="#spatialfeatureexperiment"></a>
</h2>
<p><code>SpatialFeatureExperiment</code> (SFE) is a new <a href="https://adv-r.hadley.nz/s4.html" class="external-link">S4</a> class built on top of <a href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link"><code>SpatialExperiment</code></a>
(SPE). SFE incorporates geometries and geometric operations with the <a href="https://cran.r-project.org/web/packages/sf/index.html" class="external-link"><code>sf</code></a>
package. Examples of supported geometries are Visium spots represented
with polygons corresponding to their size, cell or nuclei segmentation
polygons, tissue boundary polygons, pathologist annotation of
histological regions, and transcript spots of genes. Using
<code>sf</code>, <code>SpatialFeatureExperiment</code> leverages the
GEOS C++ library underlying <code>sf</code> for geometry operations,
including algorithms for for determining whether geometries intersect,
finding intersection geometries, buffering geometries with margins, etc.
A schematic of the SFE object is shown below:</p>
<p><img src="sfe_schematics.png" alt="SpatialFeatureExperiment expands on SpatialExperiment by adding column, row, and annotation geometries and spatial graphs. This is explained in detail in the following paragraphs." width="100%"></p>
<p>Below is a list of SFE features that extend the SPE object:</p>
<ul>
<li>
<code>colGeometries</code> are <code>sf</code> data frames
associated with the entities that correspond to columns of the gene
count matrix, such as Visium spots or cells. The geometries in the
<code>sf</code> data frames can be Visium spot centroids, Visium spot
polygons, or for datasets with single cell resolution, cell or nuclei
segmentations. Multiple <code>colGeometries</code> can be stored in the
same SFE object, such as one for cell segmentation and another for
nuclei segmentation. There can be non-spatial, attribute columns in a
<code>colGeometry</code> rather than <code>colData</code>, because the
<code>sf</code> class allows users to specify how attributes relate to
geometries, such as “constant”, “aggregate”, and “identity”. See the
<code>agr</code> argument of the <a href="https://r-spatial.github.io/sf/reference/sf.html" class="external-link"><code>st_sf</code>
documentation</a>.</li>
<li>
<code>colGraphs</code> are spatial neighborhood graphs of cells or
spots. The graphs have class <code>listw</code> (<code>spdep</code>
package), and the <code>colPairs</code> of
<code>SingleCellExperiment</code> was not used so no conversion is
necessary to use the numerous spatial dependency functions from
<code>spdep</code>, such as those for Moran’s I, Geary’s C, Getis-Ord
Gi*, LOSH, etc. Conversion is also not needed for other classical
spatial statistics packages such as <code>spatialreg</code> and
<code>adespatial</code>.</li>
<li>
<code>rowGeometries</code> are similar to
<code>colGeometries</code>, but support entities that correspond to rows
of the gene count matrix, such as genes. As we shall see below, a use
case is to store transcript spots for each gene in smFISH or in situ
sequencing based datasets.</li>
<li>
<code>rowGraphs</code> are similar to <code>colGraphs</code>. A
potential use case may be spatial colocalization of transcripts of
different genes.</li>
<li>
<code>annotGeometries</code> are <code>sf</code> data frames
associated with the dataset but not directly with the gene count matrix,
such as tissue boundaries, histological regions, cell or nuclei
segmentation in Visium datasets, and etc. These geometries are stored in
this object to facilitate plotting and using <code>sf</code> for
operations such as to find the number of nuclei in each Visium spot and
which histological regions each Visium spot intersects. Unlike
<code>colGeometries</code> and <code>rowGeometries</code>, the number of
rows in the <code>sf</code> data frames in <code>annotGeometries</code>
is not constrained by the dimension of the gene count matrix and can be
arbitrary.</li>
<li>
<code>annotGraphs</code> are similar to <code>colGraphs</code> and
<code>rowGraphs</code>, but are for entities not directly associated
with the gene count matrix, such as spatial neighborhood graphs for
nuclei in Visium datasets, or other objects like myofibers. These graphs
are relevant to <code>spdep</code> analyses of attributes of these
geometries such as spatial autocorrelation in morphological metrics of
myofibers and nuclei. With geometry operations with <code>sf</code>,
these attributes and results of analyses of these attributes
(e.g. spatial regions defined by the attributes) may be related back to
gene expression.</li>
<li>
<code>localResults</code> are similar to <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations" class="external-link"><code>reducedDims</code>
in <code>SingleCellExperiment</code></a>, but stores results from
univariate and bivariate local spatial analysis results, such as from <a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link"><code>localmoran</code></a>,
<a href="https://r-spatial.github.io/spdep/reference/localG.html" class="external-link">Getis-Ord
Gi*</a>, and <a href="https://r-spatial.github.io/spdep/reference/LOSH.html" class="external-link">local
spatial heteroscedasticity (LOSH)</a>. Unlike in
<code>reducedDims</code>, for each type of results (type is the type of
analysis such as Getis-Ord Gi*), each feature (e.g. gene) or pair of
features for which the analysis is performed has its own results. The
local spatial analyses can also be performed for attributes of
<code>colGeometries</code> and <code>annotGeometries</code> in addition
to gene expression and <code>colData</code>. Results of multivariate
spatial analysis such as <a href="https://cran.r-project.org/web/packages/adespatial/vignettes/tutorial.html#multispati-analysis" class="external-link">MULTISPATI
PCA</a> can be stored in <code>reducedDims</code>.</li>
<li>
<code>imgData</code> store images associated with the dataset. This
field is inherited from SPE, but SFE has extended the image
functionalities so images are not loaded into memory unless
necessary.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/EBImage" class="external-link">EBImage</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-an-sfe-object">Create an SFE object<a class="anchor" aria-label="anchor" href="#create-an-sfe-object"></a>
</h2>
<div class="section level3">
<h3 id="visium">Visium<a class="anchor" aria-label="anchor" href="#visium"></a>
</h3>
<p>10x Genomics Space Ranger output from a Visium experiment can be read
in a similar manner as in <code>SpatialExperiment</code>; the
<code>SpatialFeatureExperiment</code> SFE object has the
<code>spotPoly</code> column geometry for the spot polygons. If the
filtered matrix (i.e. only spots in the tissue) is read in, then a
column graph called <code>visium</code> will also be present for the
spatial neighborhood graph of the Visium spots on the tissue. The graph
is not computed if all spots are read in regardless of whether they are
on tissue.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample01"</span>, <span class="st">"sample02"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/usr/local/lib/R/site-library/SpatialFeatureExperiment/extdata/sample01"</span></span>
<span><span class="co">#&gt; [2] "/usr/local/lib/R/site-library/SpatialFeatureExperiment/extdata/sample02"</span></span></code></pre></div>
<p>The results for each tissue capture should be in the
<code>outs</code> directory under the sample directory. Inside the
<code>outs</code> directory, these directories may be present:
<code>raw_reature_bc_matrix</code> has the unfiltered gene count matrix,
<code>filtered_feature_bc_matrix</code> has the gene count matrix for
spots in tissue, and <code>spatial</code> has the spatial information.
The matrix directories contain the matrices in MTX format as sparse
matrices. Space Ranger also outputs the matrices as h5 files, which are
read into R in a similar way as MTX.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "filtered_feature_bc_matrix" "spatial"</span></span></code></pre></div>
<p>Inside the matrix directory:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span>, <span class="st">"filtered_feature_bc_matrix"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcodes.tsv" "features.tsv" "matrix.mtx"</span></span></code></pre></div>
<p>Inside the <code>spatial</code> directory:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"outs"</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "aligned_fiducials.jpg"              "barcode_fluorescence_intensity.csv"</span></span>
<span><span class="co">#&gt; [3] "detected_tissue_image.jpg"          "scalefactors_json.json"            </span></span>
<span><span class="co">#&gt; [5] "spatial_enrichment.csv"             "tissue_hires_image.png"            </span></span>
<span><span class="co">#&gt; [7] "tissue_lowres_image.png"            "tissue_positions.csv"</span></span></code></pre></div>
<p><code>tissue_lowres_image.png</code> is a low resolution image of the
tissue. Not all Visium datasets have all the files here. The
<code>barcode_fluorescence_intensity.csv</code> file is only present in
datasets with fluorescent imaging rather than bright field H&amp;E.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, sample_id <span class="op">=</span> <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, </span>
<span>                          data <span class="op">=</span> <span class="st">"filtered"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 25 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ENSG00000014257 ENSG00000142515 ENSG00000263639</span></span>
<span><span class="co">#&gt;   ENSG00000163810 ENSG00000149591</span></span>
<span><span class="co">#&gt; rowData names(14): symbol Feature.Type ...</span></span>
<span><span class="co">#&gt;   Median.Normalized.Average.Counts_sample02</span></span>
<span><span class="co">#&gt;   Barcodes.Detected.per.Feature_sample02</span></span>
<span><span class="co">#&gt; colnames(25): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   TGCAATTTGGGCACGG-1 ATGCCAATCGCTCTGC-1</span></span>
<span><span class="co">#&gt; colData names(10): in_tissue array_row ... channel3_mean channel3_stdev</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01: col: visium</span></span>
<span><span class="co">#&gt; sample02: col: visium</span></span></code></pre></div>
<p>Space Ranger output includes the gene count matrix, spot coordinates,
and spot diameter. The Space Ranger output does NOT include nuclei
segmentation or pathologist annotation of histological regions. Extra
image processing, such as with ImageJ and QuPath, are required for those
geometries.</p>
</div>
<div class="section level3">
<h3 id="vizgen-merfish">Vizgen MERFISH<a class="anchor" aria-label="anchor" href="#vizgen-merfish"></a>
</h3>
<p>As of Bioc 3.19, the following read functions have been implemented
in SFE: <code>read10xVisiumSFE</code>, <code>readVizgen</code>
(MERFISH), <code>readCosMX</code>, and <code>readXenium</code>. Read
functions for more commercial technologies will be implemented soon, and
you can help writing one in this hackathon, such as for <a href="https://github.com/pachterlab/SpatialFeatureExperiment/issues/18" class="external-link">Molecular
Cartography</a>. Here we demonstrate MERFISH</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/vizgen_cellbound"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">fn</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cell_boundaries"          "cell_boundaries.parquet" </span></span>
<span><span class="co">#&gt; [3] "cell_by_gene.csv"         "cell_metadata.csv"       </span></span>
<span><span class="co">#&gt; [5] "detected_transcripts.csv" "images"</span></span></code></pre></div>
<p>The <code>cell_boundaries</code> directory has <code>hdf5</code> file
with cell segmentation polygons, which are rather slow to read, so it’s
faster to read the cell segmentation polygons from the
<code>cell_boundaries.parquet</code> file. The <a href="https://www.databricks.com/glossary/what-is-parquet" class="external-link">parquet
file</a> stores column-oriented data more efficiently than old fashioned
<code>csv</code> and can contain Simple Feature geometry data. The
<code>cell_by_gene.csv</code> file is the gene count matrix.
<code>detected_transcripts.csv</code> has coordinates of transcript
spots and is reformatted into a new parquet file when
<code>readVizgen</code> is called with <code>add_molecules = TRUE</code>
(default is <code>FALSE</code>) so the transcript spots can be read into
R much more quickly next time the data is read. For larger datasets,
this reformatting can take a while.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="va">fn</span>, <span class="st">"."</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="op">(</span><span class="va">sfe_vizgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen</a></span><span class="op">(</span><span class="st">"vizgen_cellbound"</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 1 `.parquet` files exist:</span></span>
<span><span class="co">#&gt; /__w/VoyagerHackathon/VoyagerHackathon/vignettes/vizgen_cellbound/cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; using -&gt; /__w/VoyagerHackathon/VoyagerHackathon/vignettes/vizgen_cellbound/cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file</span></span>
<span><span class="co">#&gt; Removing 35 cells with area less than 15</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering geometries to match 1023 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Total of 26 features/genes with `min_phred` &lt; 20 are removed from SFE object</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 62 1023 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(62): CD4 TLL1 ... Blank-37 Blank-39</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1023): 112824700230101267 112824700230101269 ...</span></span>
<span><span class="co">#&gt;   112824700330100848 112824700330100920</span></span>
<span><span class="co">#&gt; colData names(11): fov volume ... solidity sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>Look at the image data stored in the <code>imgData</code> field; the
actual images are in the data column. We will go over the getters of the
images and plot the images later.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData</a></span><span class="op">(</span><span class="va">sfe_vizgen</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 5 rows and 4 columns</span></span>
<span><span class="co">#&gt;     sample_id      image_id   data scaleFactor</span></span>
<span><span class="co">#&gt;   &lt;character&gt;   &lt;character&gt; &lt;list&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1    sample01 Cellbound1_z3   ####           1</span></span>
<span><span class="co">#&gt; 2    sample01 Cellbound2_z3   ####           1</span></span>
<span><span class="co">#&gt; 3    sample01 Cellbound3_z3   ####           1</span></span>
<span><span class="co">#&gt; 4    sample01       DAPI_z3   ####           1</span></span>
<span><span class="co">#&gt; 5    sample01      PolyT_z3   ####           1</span></span></code></pre></div>
<p>The <code>images</code> directory contains images from 3 types of
cell membrane stains, DAPI (nuclei), and PolyT (all polyadenylated
transcripts, visualizing whole cells). When there are multiple z-planes,
by default the center plane (the 3rd plane) is read, but this can be
changed. The cell segmentation is the same in all z-planes from Vizgen
output.</p>
<p>See <a href="https://pachterlab.github.io/voyager/articles/create_sfe.html" class="external-link">this
vignette</a> on creating SFE objects from scratch and for other spatial
trancriptomics technologies.</p>
</div>
</div>
<div class="section level2">
<h2 id="non-spatial-operations-of-sfe-objects">Non-spatial operations of SFE objects<a class="anchor" aria-label="anchor" href="#non-spatial-operations-of-sfe-objects"></a>
</h2>
<p>Operations on SFE objects are demonstrated on a small toy dataset
(you may need to answer a prompt in the R console when downloading the
dataset):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 77 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(77): AAATTACCTATCGATG AACATATCAACTGGTG ... TTCTTTGGTCGCGACG</span></span>
<span><span class="co">#&gt;   TTGATGTGTAGTCCCG</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<div class="section level3">
<h3 id="singlecellexperiment-getters-and-setters">
<code>SingleCellExperiment</code> getters and setters<a class="anchor" aria-label="anchor" href="#singlecellexperiment-getters-and-setters"></a>
</h3>
<p>Each SFE object is an SCE object as SFE builds on top of SCE, or
inherits from SCE, so all the SCE methods apply. Here “inherits” is just
like each bioinformatician is a human, where “bioinformatician” is a bit
like SFE and “human” is like SCE. Here we go over SCE getters and
setters.</p>
<p>At the center of SCE is the gene count matrix. You can get or set the
gene count matrix with <code>counts</code> function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="co">#&gt; 6 x 77 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;   [[ suppressing 77 column names 'AAATTACCTATCGATG', 'AACATATCAACTGGTG', 'AAGATTGGCGGAACGT' ... ]]</span></span>
<span><span class="co">#&gt;                                                                               </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 . . . . . . . . . . . . . . . . . 2 1 . . . . . . 1 . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 1 1 2 4 1 1 . . . . . 1 . . . . . 1 2 . 1 . . 2 2 1 3 . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 . . 1 . . . . . . . . . . . . . . . . . 1 . 1 1 . 1 2 3 . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 . . . . . . . . . 1 . . . . . . . 2 . . . . 2 . 1 1 . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000002459 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .</span></span>
<span><span class="co">#&gt;                                                                               </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 . . . 1 . . . . . . . . . 1 . . . . . . . . . . . . . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 . . . 2 . . 1 . 1 . . 2 . 1 . . 1 . 1 1 . . 1 1 1 . . 1 . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 1 . . 1 . 1 . . 2 . . 1 1 . . . . . . . 1 . . 1 1 3 . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 . . . . 1 . . . . . . . . . . 1 . . 1 1 1 . . . . 1 . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000002459 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .</span></span>
<span><span class="co">#&gt;                                                     </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 . . . 1 . . 2 1 . . . . . . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126 . . . . . . . . . . . . . . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 . . . 3 . . 1 . . . . . 2 2 . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 . . . . . . . . . . . . . . . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 . . . . . . . . . . . . . 1 . . .</span></span>
<span><span class="co">#&gt; ENSMUSG00000002459 . . . . . . . . . . . . . . . . .</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">m</span></span></code></pre></div>
<p>After log normalizing data, similarly the <code>logcounts</code>
function can be used to get or set the log normalized gene count
matrix.</p>
<p>The gene count matrix has metadata about the cells and genes. Use the
<code>colData</code> function to get cell metadata and
<code>rowData</code> to get gene metadata</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 77 rows and 12 columns</span></span>
<span><span class="co">#&gt;                           barcode       col       row         x         y</span></span>
<span><span class="co">#&gt;                       &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG AAATTACCTATCGATG        94        26      9525     41335</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG AACATATCAACTGGTG        99        27      9775     41422</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT AAGATTGGCGGAACGT        90        26      9325     41335</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT AAGGGACAGATTCTGT        92        26      9425     41335</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC AATATCGAGGGTTCTC       101        27      9875     41422</span></span>
<span><span class="co">#&gt; ...                           ...       ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; TTAGCTAATACGATCT TTAGCTAATACGATCT        98        22      9725     40987</span></span>
<span><span class="co">#&gt; TTCCAATCAGAGCTAG TTCCAATCAGAGCTAG        95        23      9575     41074</span></span>
<span><span class="co">#&gt; TTCCGCAGAGAAATAT TTCCGCAGAGAAATAT       101        21      9875     40900</span></span>
<span><span class="co">#&gt; TTCTTTGGTCGCGACG TTCTTTGGTCGCGACG       104        24     10025     41161</span></span>
<span><span class="co">#&gt; TTGATGTGTAGTCCCG TTGATGTGTAGTCCCG        90        24      9325     41161</span></span>
<span><span class="co">#&gt;                        dia    tissue   sample_id   nCounts    nGenes prop_mito</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt; &lt;logical&gt; &lt;character&gt; &lt;numeric&gt; &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG   180.278      TRUE       Vis5A     13229      2146  0.345680</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG   180.278      TRUE       Vis5A     10197      1782  0.411297</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT   180.278      TRUE       Vis5A      9597      1769  0.383036</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT   180.278      TRUE       Vis5A     15855      2355  0.383728</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC   180.278      TRUE       Vis5A     10327      1683  0.521061</span></span>
<span><span class="co">#&gt; ...                    ...       ...         ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; TTAGCTAATACGATCT   180.278      TRUE       Vis5A      1736       625  0.311060</span></span>
<span><span class="co">#&gt; TTCCAATCAGAGCTAG   180.278      TRUE       Vis5A     13706      2696  0.338903</span></span>
<span><span class="co">#&gt; TTCCGCAGAGAAATAT   180.278        NA       Vis5A       115        73  0.304348</span></span>
<span><span class="co">#&gt; TTCTTTGGTCGCGACG   180.278        NA       Vis5A        96        56  0.437500</span></span>
<span><span class="co">#&gt; TTGATGTGTAGTCCCG   180.278      TRUE       Vis5A      9948      1808  0.458786</span></span>
<span><span class="co">#&gt;                  in_tissue</span></span>
<span><span class="co">#&gt;                  &lt;logical&gt;</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG      TRUE</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG      TRUE</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT      TRUE</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT      TRUE</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC      TRUE</span></span>
<span><span class="co">#&gt; ...                    ...</span></span>
<span><span class="co">#&gt; TTAGCTAATACGATCT      TRUE</span></span>
<span><span class="co">#&gt; TTCCAATCAGAGCTAG      TRUE</span></span>
<span><span class="co">#&gt; TTCCGCAGAGAAATAT     FALSE</span></span>
<span><span class="co">#&gt; TTCTTTGGTCGCGACG     FALSE</span></span>
<span><span class="co">#&gt; TTGATGTGTAGTCCCG      TRUE</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 15123 rows and 6 columns</span></span>
<span><span class="co">#&gt;                               Ensembl      symbol            type       means</span></span>
<span><span class="co">#&gt;                           &lt;character&gt; &lt;character&gt;     &lt;character&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 ENSMUSG00000025902       Sox17 Gene Expression 0.007612179</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126 ENSMUSG00000096126     Gm22307 Gene Expression 0.000200321</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 ENSMUSG00000033845      Mrpl15 Gene Expression 0.075921474</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 ENSMUSG00000025903      Lypla1 Gene Expression 0.057491987</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 ENSMUSG00000033813       Tcea1 Gene Expression 0.052283654</span></span>
<span><span class="co">#&gt; ...                               ...         ...             ...         ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000064360 ENSMUSG00000064360      mt-Nd3 Gene Expression  11.3030849</span></span>
<span><span class="co">#&gt; ENSMUSG00000064363 ENSMUSG00000064363      mt-Nd4 Gene Expression  24.8173077</span></span>
<span><span class="co">#&gt; ENSMUSG00000064367 ENSMUSG00000064367      mt-Nd5 Gene Expression   2.9625401</span></span>
<span><span class="co">#&gt; ENSMUSG00000064368 ENSMUSG00000064368      mt-Nd6 Gene Expression   0.0240385</span></span>
<span><span class="co">#&gt; ENSMUSG00000064370 ENSMUSG00000064370     mt-Cytb Gene Expression  24.2099359</span></span>
<span><span class="co">#&gt;                           vars       cv2</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 0.008757912  151.1411</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126 0.000200321 4992.0000</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 0.114250804   19.8212</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 0.080645121   24.3985</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 0.073603279   26.9256</span></span>
<span><span class="co">#&gt; ...                        ...       ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000064360 9.35532e+02   7.32259</span></span>
<span><span class="co">#&gt; ENSMUSG00000064363 5.00034e+03   8.11877</span></span>
<span><span class="co">#&gt; ENSMUSG00000064367 7.59503e+01   8.65369</span></span>
<span><span class="co">#&gt; ENSMUSG00000064368 2.98769e-02  51.70369</span></span>
<span><span class="co">#&gt; ENSMUSG00000064370 4.96740e+03   8.47505</span></span></code></pre></div>
<p>Just like in Seurat, the SCE object can be subsetted like a matrix.
Here not all Visium spots intersect the tissue. In this dataset, whether
the spot intersects tissue is in a column in <code>colData</code> called
<code>in_tissue</code>, and we’ll subset the SFE object to only keep
spots in tissue, and to only keep genes that are detected.
<code>colData</code> columns in SCE can be accessed with the
<code>$</code> operator as if getting a column from a data frame.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="va">sfe</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<p>PCA is part of the standard scRNA-seq data analysis workflow. Here
we’ll first normalize the data and then perform PCA and get the PCA
results.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co"># Log counts getter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 6 x 57 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;   [[ suppressing 57 column names 'AAATTACCTATCGATG', 'AACATATCAACTGGTG', 'AAGATTGGCGGAACGT' ... ]]</span></span>
<span><span class="co">#&gt;                                                                           </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 .         .        .        .         .       .        </span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 0.9120341 1.100212 1.778525 1.9791690 1.09049 0.7212191</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 .         .        1.147552 .         .       .        </span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 .         .        .        .         .       .        </span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 .         .        .        .         1.09049 0.7212191</span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 0.9120341 1.100212 .        0.7954883 .       1.1998445</span></span>
<span><span class="co">#&gt;                                                                          </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 .         .         .        . . 1.4235808 0.9221123 .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 .         .         1.187124 . . 0.8806874 1.4801489 .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 .         .         .        . . .         .         .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 .         0.7819597 .        . . 1.4235808 .         .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 .         .         .        . . .         .         .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 0.9664523 .         .        . . 0.8806874 0.9221123 .</span></span>
<span><span class="co">#&gt;                                                                             </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 .         .         .         .        0.7434199 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 0.7597265 .         1.5084926 1.626531 0.7434199 1.756849</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 0.7597265 0.8244546 0.9430309 .        0.7434199 1.370942</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 .         1.3458113 .         1.031288 0.7434199 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 .         .         0.9430309 .        .         .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 0.7597265 0.8244546 .         1.031288 0.7434199 1.370942</span></span>
<span><span class="co">#&gt;                                                                            </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 .        . .        .         0.8901568 .        .      </span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 .        . .        .         1.4365644 .        .      </span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 2.255916 . 1.284466 .         0.8901568 .        1.79022</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 .        . .        .         .         1.616818 .      </span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 .        . .        .         .         .        .      </span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 1.175551 . 1.953018 0.9455567 0.8901568 .        .      </span></span>
<span><span class="co">#&gt;                                                                               </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 .         .        .         . .         .         1.755744</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 0.8038305 .        0.7911927 . 1.3328186 .         1.755744</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 .         .        1.2992499 . 0.8151422 0.8999878 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 .         .        .         . .         .         .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 .         .        .         . .         .         .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 2.2414286 1.093611 0.7911927 . .         .         .       </span></span>
<span><span class="co">#&gt;                                                                          </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 .    .        . .         .         .         .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 .    1.452596 . 0.7671356 0.8750204 .         1.711303</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 .    .        . .         .         0.8493164 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 1.08 .        . 0.7671356 0.8750204 0.8493164 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 1.08 .        . .         .         .         .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 .    .        . .         0.8750204 .         1.711303</span></span>
<span><span class="co">#&gt;                                                                               </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 .        .         .         .         .         .        .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 1.158464 0.8717957 .         .         0.6711848 .        .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 1.158464 0.8717957 1.7921684 .         .         .        .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 .        .         0.8648235 .         .         .        .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 .        .         .         .         .         .        .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 .        .         2.0991159 0.8960287 0.6711848 2.119829 .</span></span>
<span><span class="co">#&gt;                                                                                </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 0.9283765 . 1.4077190 1.35168 .         . .        .        </span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 1.8911857 . 0.8691502 .       .         . 3.851774 1.4340401</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 .         . .         .       .         . .        .        </span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 .         . .         .       .         . .        0.8883139</span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 .         . .         .       .         . .        0.8883139</span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 .         . 1.4077190 1.35168 0.8601784 . .        1.4340401</span></span>
<span><span class="co">#&gt;                            </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000033793 .       </span></span>
<span><span class="co">#&gt; ENSMUSG00000025907 1.119352</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Highly variable genes</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">10</span>, subset_row <span class="op">=</span> <span class="va">hvgs</span>,</span>
<span>                     exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Later we will see that Voyager spatial analysis functions are modeled
after <code>runPCA</code> un user interface. The <code>reducedDim</code>
function can be used to get and set dimension reduction results. User
interfaces to get or set the geometries and spatial graphs emulate those
of <code>reducedDims</code> and <code>row/colPairs</code> in
<code>SingleCellExperiment</code>. Column and row geometries also
emulate <code>reducedDims</code> in internal implementation, while
annotation geometries and spatial graphs differ.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pca_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         PC1        PC2         PC3        PC4        PC5</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG  1.3377315 -0.6272455 -1.89418096 -0.4281142  1.6679033</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG  0.9176196 -1.7446527 -1.17466111 -0.1101124 -0.8328961</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT -0.3127939 -2.3381292  1.55334449  2.0582206  3.1237670</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT  1.1454887 -1.6483366 -1.34405438 -0.6489710  2.0769778</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC  1.6871434 -1.3772805 -2.99001101 -0.7206788  1.4824973</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT  0.9398691 -0.3118155  0.02960312  0.7916621  0.9375853</span></span>
<span><span class="co">#&gt;                         PC6       PC7        PC8        PC9       PC10</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG -1.7815613 0.3881873 -2.8266465  4.1393556  1.5509855</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG  0.3579236 3.1559300 -0.2434049  0.5127298  0.3363097</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT -3.9268637 2.4060020 -1.2217007  3.1234996 -1.3912454</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT -0.5549159 2.4213310 -1.5162314  0.6479754  0.6833253</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC -0.3287834 6.1201618  5.0054816 -4.9051787 -3.9949746</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT  0.7241128 1.7345248  1.0288884 -1.2416471  1.3317063</span></span></code></pre></div>
<p>Here the second argument is used to specify which dimension reduction
result to get. If it’s not specified, then by default the first one is
retrieved, so the code below would be equivalent to the previous
chunk:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         PC1        PC2         PC3        PC4        PC5</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG  1.3377315 -0.6272455 -1.89418096 -0.4281142  1.6679033</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG  0.9176196 -1.7446527 -1.17466111 -0.1101124 -0.8328961</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT -0.3127939 -2.3381292  1.55334449  2.0582206  3.1237670</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT  1.1454887 -1.6483366 -1.34405438 -0.6489710  2.0769778</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC  1.6871434 -1.3772805 -2.99001101 -0.7206788  1.4824973</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT  0.9398691 -0.3118155  0.02960312  0.7916621  0.9375853</span></span>
<span><span class="co">#&gt;                         PC6       PC7        PC8        PC9       PC10</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG -1.7815613 0.3881873 -2.8266465  4.1393556  1.5509855</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG  0.3579236 3.1559300 -0.2434049  0.5127298  0.3363097</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT -3.9268637 2.4060020 -1.2217007  3.1234996 -1.3912454</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT -0.5549159 2.4213310 -1.5162314  0.6479754  0.6833253</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC -0.3287834 6.1201618  5.0054816 -4.9051787 -3.9949746</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT  0.7241128 1.7345248  1.0288884 -1.2416471  1.3317063</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set PCA embeddings say if you ran PCA elsewhere</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pca_res</span></span></code></pre></div>
<p>Which dimension reductions are present?</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "PCA"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="column-geometries">Column geometries<a class="anchor" aria-label="anchor" href="#column-geometries"></a>
</h3>
<p>Column geometries or <code>colGeometries</code> are the geometries
that correspond to columns of the gene count matrix, such as Visium
spots and cells in datasets from a single cell resolution technology.
Each SFE object can have multiple column geometries. For example, in a
dataset with single cell resolution, whole cell segmentation and nuclei
segmentation are two different <code>colGeometries</code>. However, for
Visium, the spot polygons are the only <code>colGeometry</code>
obviously relevant, though users can add other geometries such as
results of geometric operations on the spot polygons. The different
geometries can be get or set with their names, and “spotPoly” is the
standard name for Visium spot polygons.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get Visium spot polygons</span></span>
<span><span class="op">(</span><span class="va">spots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"spotPoly"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 57 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5000 ymin: 13000 xmax: 7000 ymax: 14931.74</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                                        geometry sample_id</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG POLYGON ((6472.186 13875.23...     Vis5A</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG POLYGON ((5778.291 13635.43...     Vis5A</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT POLYGON ((7000 13809.84, 69...     Vis5A</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT POLYGON ((6749.535 13874.64...     Vis5A</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC POLYGON ((5500.941 13636.03...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT POLYGON ((6612.42 14598.82,...     Vis5A</span></span>
<span><span class="co">#&gt; AATTCATAAGGGATCT POLYGON ((6889.769 14598.22...     Vis5A</span></span>
<span><span class="co">#&gt; ACGCTAGTGATACACT POLYGON ((5639.096 13394.44...     Vis5A</span></span>
<span><span class="co">#&gt; AGCGCGGGTGCCAATG POLYGON ((6750.575 14357.23...     Vis5A</span></span>
<span><span class="co">#&gt; AGTGAGCCTCGCCGCC POLYGON ((5222.552 13154.04...     Vis5A</span></span></code></pre></div>
<p>Here we get a <code>sf</code> data frame, which is just like a
regular data frame but with a special <code>geometry</code> column. Now
plot these spot polygons</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">spots</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set colGeometry</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots</span></span></code></pre></div>
<p>To see which <code>colGeometries</code> are present in the SFE
object:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometryNames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "spotPoly"</span></span></code></pre></div>
<p>There are shorthands for some specific column or row geometries. For
example, <code>spotPoly(sfe)</code> is equivalent to
<code>colGeometry(sfe, "spotPoly")</code> shown above.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">spots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 57 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5000 ymin: 13000 xmax: 7000 ymax: 14931.74</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                                        geometry sample_id</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG POLYGON ((6472.186 13875.23...     Vis5A</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG POLYGON ((5778.291 13635.43...     Vis5A</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT POLYGON ((7000 13809.84, 69...     Vis5A</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT POLYGON ((6749.535 13874.64...     Vis5A</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC POLYGON ((5500.941 13636.03...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT POLYGON ((6612.42 14598.82,...     Vis5A</span></span>
<span><span class="co">#&gt; AATTCATAAGGGATCT POLYGON ((6889.769 14598.22...     Vis5A</span></span>
<span><span class="co">#&gt; ACGCTAGTGATACACT POLYGON ((5639.096 13394.44...     Vis5A</span></span>
<span><span class="co">#&gt; AGCGCGGGTGCCAATG POLYGON ((6750.575 14357.23...     Vis5A</span></span>
<span><span class="co">#&gt; AGTGAGCCTCGCCGCC POLYGON ((5222.552 13154.04...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots</span></span></code></pre></div>
<p>Exercise: The <code>cellSeg</code> function gets cell segmentation
from the MERFISH dataset in <code>sfe_vizgen</code>. Get and plot the
cell segmentations.</p>
</div>
<div class="section level3">
<h3 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h3>
<p>Annotation geometries can be get or set with
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry()</a></code>. In column or row geometries, the number of
rows of the <code>sf</code> data frame (i.e. the number of geometries in
the data frame) is constrained by the number of rows or columns of the
gene count matrix respectively, because just like <code>rowData</code>
and <code>colData</code>, each row of a <code>rowGeometry</code> or
<code>colGeometry</code> <code>sf</code> data frame must correspond to a
row or column of the gene count matrix respectively. In contrast, an
<code>annotGeometry</code> <code>sf</code> data frame can have any
dimension, not constrained by the dimension of the gene count
matrix.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter, by name or index</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"tissueBoundary"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 1 feature and 2 fields</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5094 ymin: 13000 xmax: 7000 ymax: 14969</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt;   ID                       geometry sample_id</span></span>
<span><span class="co">#&gt; 7  7 POLYGON ((5094 13000, 5095 ...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tb</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter, by name or index</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span></span></code></pre></div>
<p>See which <code>annotGeometries</code> are present in the SFE
object:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "tissueBoundary"      "myofiber_full"       "myofiber_simplified"</span></span>
<span><span class="co">#&gt; [4] "nuclei"              "nuclei_centroid"</span></span></code></pre></div>
<p>There are shorthands for specific annotation geometries. For example,
<code>tissueBoundary(sfe)</code> is equivalent to
<code>annotGeometry(sfe, "tissueBoundary")</code>.
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">cellSeg()</a></code> (cell segmentation) and <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">nucSeg()</a></code>
(nuclei segmentation) would first query <code>colGeometries</code> (for
single cell, single molecule technologies, equivalent to
<code>colGeometry(sfe, "cellSeg")</code> or
<code>colGeometry(sfe, "nucSeg")</code>), and if not found, they will
query <code>annotGeometries</code> (for array capture and
microdissection technologies, equivalent to
<code>annotGeometry(sfe, "cellSeg")</code> or
<code>annotGeometry(sfe, "nucSeg")</code>).</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 1 feature and 2 fields</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5094 ymin: 13000 xmax: 7000 ymax: 14969</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt;   ID                       geometry sample_id</span></span>
<span><span class="co">#&gt; 7  7 POLYGON ((5094 13000, 5095 ...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="spatial-graphs">Spatial graphs<a class="anchor" aria-label="anchor" href="#spatial-graphs"></a>
</h3>
<p>The spatial neighborhood graphs for Visium spots are stored in the
<code>colGraphs</code> field, which has similar user interface as
<code>colGeometries</code>. SFE also wraps all methods to find the
spatial neighborhood graph implemented in the <code>spdep</code>
package, and triangulation is used here as demonstration.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">"tri2nb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 57 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 314 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 9.664512 </span></span>
<span><span class="co">#&gt; Average number of links: 5.508772 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 57 3249 57 20.93315 229.9213</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">g</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph by name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"graph1"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">g</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get graph by name</span></span>
<span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"graph1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 57 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 314 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 9.664512 </span></span>
<span><span class="co">#&gt; Average number of links: 5.508772 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 57 3249 57 20.93315 229.9213</span></span></code></pre></div>
<p>For Visium, spatial neighborhood graph of the hexagonal grid can be
found with the known locations of the barcodes. One SFE object can have
multiple <code>colGraphs</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-41-1.png" width="700"></p>
<p>Which graphs are present in this SFE object?</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraphNames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "graph1" "visium"</span></span></code></pre></div>
<p>While this workshop only works with one sample, i.e. tissue section,
operations on multiple samples is discussed in <a href="https://pachterlab.github.io/SpatialFeatureExperiment/articles/SFE.html#multiple-samples" class="external-link">the
vignette of the SFE package</a>.</p>
</div>
<div class="section level3">
<h3 id="row-geometries">Row geometries<a class="anchor" aria-label="anchor" href="#row-geometries"></a>
</h3>
<p>The Visium dataset does not have geometries associated with genes,
but the MERFISH dataset does. The <code>rowGeometry</code> getter and
setter have pretty much the same user interface as the getters and
setters covered above:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">rg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/rowGeometries.html" class="external-link">rowGeometry</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, <span class="st">"txSpots"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 62 features and 4 fields</span></span>
<span><span class="co">#&gt; Geometry type: MULTIPOINT</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 6500.184 ymin: -1499.778 xmax: 6799.719 ymax: -1200.041</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;          gene barcode_id global_z   transcript_id</span></span>
<span><span class="co">#&gt; CD4       CD4          0        3 ENST00000011653</span></span>
<span><span class="co">#&gt; TLL1     TLL1          1        3 ENST00000061240</span></span>
<span><span class="co">#&gt; EPHA8   EPHA8          2        3 ENST00000166244</span></span>
<span><span class="co">#&gt; CDKN1A CDKN1A          4        3 ENST00000244741</span></span>
<span><span class="co">#&gt; PDGFRA PDGFRA          6        3 ENST00000257290</span></span>
<span><span class="co">#&gt; CD44     CD44          7        3 ENST00000263398</span></span>
<span><span class="co">#&gt; FCGR2A FCGR2A          8        3 ENST00000271450</span></span>
<span><span class="co">#&gt; EGFR     EGFR          9        3 ENST00000275493</span></span>
<span><span class="co">#&gt; DRAXIN DRAXIN         11        3 ENST00000294485</span></span>
<span><span class="co">#&gt; LMX1A   LMX1A         12        3 ENST00000294816</span></span>
<span><span class="co">#&gt;                              geometry</span></span>
<span><span class="co">#&gt; CD4    MULTIPOINT ((6770.52 -1466....</span></span>
<span><span class="co">#&gt; TLL1   MULTIPOINT ((6692.258 -1435...</span></span>
<span><span class="co">#&gt; EPHA8  MULTIPOINT ((6531.73 -1427....</span></span>
<span><span class="co">#&gt; CDKN1A MULTIPOINT ((6774.421 -1477...</span></span>
<span><span class="co">#&gt; PDGFRA MULTIPOINT ((6642.847 -1355...</span></span>
<span><span class="co">#&gt; CD44   MULTIPOINT ((6530.448 -1449...</span></span>
<span><span class="co">#&gt; FCGR2A MULTIPOINT ((6589.974 -1481...</span></span>
<span><span class="co">#&gt; EGFR   MULTIPOINT ((6655.547 -1419...</span></span>
<span><span class="co">#&gt; DRAXIN MULTIPOINT ((6708.042 -1317...</span></span>
<span><span class="co">#&gt; LMX1A  MULTIPOINT ((6797.345 -1240...</span></span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/rowGeometries.html" class="external-link">rowGeometry</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, <span class="st">"txSpots"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rg</span></span></code></pre></div>
<p>In the case of transcript spots, there’s a special convenience
function <code>txSpots</code></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/rowGeometries.html" class="external-link">txSpots</a></span><span class="op">(</span><span class="va">sfe_vizgen</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 62 features and 4 fields</span></span>
<span><span class="co">#&gt; Geometry type: MULTIPOINT</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 6500.184 ymin: -1499.778 xmax: 6799.719 ymax: -1200.041</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;          gene barcode_id global_z   transcript_id</span></span>
<span><span class="co">#&gt; CD4       CD4          0        3 ENST00000011653</span></span>
<span><span class="co">#&gt; TLL1     TLL1          1        3 ENST00000061240</span></span>
<span><span class="co">#&gt; EPHA8   EPHA8          2        3 ENST00000166244</span></span>
<span><span class="co">#&gt; CDKN1A CDKN1A          4        3 ENST00000244741</span></span>
<span><span class="co">#&gt; PDGFRA PDGFRA          6        3 ENST00000257290</span></span>
<span><span class="co">#&gt; CD44     CD44          7        3 ENST00000263398</span></span>
<span><span class="co">#&gt; FCGR2A FCGR2A          8        3 ENST00000271450</span></span>
<span><span class="co">#&gt; EGFR     EGFR          9        3 ENST00000275493</span></span>
<span><span class="co">#&gt; DRAXIN DRAXIN         11        3 ENST00000294485</span></span>
<span><span class="co">#&gt; LMX1A   LMX1A         12        3 ENST00000294816</span></span>
<span><span class="co">#&gt;                              geometry</span></span>
<span><span class="co">#&gt; CD4    MULTIPOINT ((6770.52 -1466....</span></span>
<span><span class="co">#&gt; TLL1   MULTIPOINT ((6692.258 -1435...</span></span>
<span><span class="co">#&gt; EPHA8  MULTIPOINT ((6531.73 -1427....</span></span>
<span><span class="co">#&gt; CDKN1A MULTIPOINT ((6774.421 -1477...</span></span>
<span><span class="co">#&gt; PDGFRA MULTIPOINT ((6642.847 -1355...</span></span>
<span><span class="co">#&gt; CD44   MULTIPOINT ((6530.448 -1449...</span></span>
<span><span class="co">#&gt; FCGR2A MULTIPOINT ((6589.974 -1481...</span></span>
<span><span class="co">#&gt; EGFR   MULTIPOINT ((6655.547 -1419...</span></span>
<span><span class="co">#&gt; DRAXIN MULTIPOINT ((6708.042 -1317...</span></span>
<span><span class="co">#&gt; LMX1A  MULTIPOINT ((6797.345 -1240...</span></span></code></pre></div>
<p>Plot the transcript spots (for the toy dataset, they were subsampled
to keep the SFE package small; usually they are much denser)</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">cellSeg</a></span><span class="op">(</span><span class="va">sfe_vizgen</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rg</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="images">Images<a class="anchor" aria-label="anchor" href="#images"></a>
</h3>
<p>In SPE, the images are only used for visualization, but SFE extended
the SPE image functionality so large images don’t have to be loaded into
memory unless necessary. In SFE, there are 3 types of images:</p>
<ol style="list-style-type: decimal">
<li>
<code>SpatRasterImage</code>, the default, a thin wrapper around the
<a href="https://rspatial.github.io/terra/reference/SpatRaster-class.html" class="external-link"><code>SpatRaster</code></a>
class in the <code>terra</code> object to make it conform to SPE’s
requirements. Large images are not loaded into memory unless necessary
and it’s possible to only load a down sampled lower resolution version
of the image into memory. Spatial extent is part of
<code>SpatRaster</code>. The extent is important to delineate where the
image is in the coordinate system within the tissue section. This is a
more sophisticated way to make sure the image is aligned with the
geometries than the scale factor in SPE which only works for Visium and
would not allow the SPE object to be cropped.</li>
<li>
<code>BioFormatsImage</code> is used for <a href="https://docs.openmicroscopy.org/ome-model/5.6.3/ome-tiff/" class="external-link"><code>OME-TIFF</code></a>
images whose compression can’t be read by <code>terra</code>. The image
is not loaded into memory. It’s just some metadata, which includes the
file path, extent, and origin (minimum value of coordinates). So far
functions related to <code>BioFormatsImage</code> cater to Xenium
data.</li>
<li>
<code>EBImage</code> is a thin wrapper around the <a href="https://bioconductor.org/packages/release/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html#3_Image_data_representation" class="external-link"><code>Image</code></a>
class in the <code>EBImage</code> package to conform to SPE’s
requirements. With <code>EBImage</code>, one can do thresholding and
morphological operations. However, it’s not merely a wrapper; it
contains another metadata field for the extent. When
<code>BioFormatsImage</code> is loaded into memory, it becomes
<code>EBImage</code>.</li>
</ol>
<p>In the MERFISH dataset above, the image is represented as
<code>SpatRasterImage</code>. Get images with the <code>getImg</code>
function, and use the <code>image_id</code> function to indicate which
image to get, and if it’s left blank, the first imaget will be
retrieved:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, image_id <span class="op">=</span> <span class="st">"DAPI_z3"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 1 x 695 x 695 (width x height) SpatRasterImage</span></span>
<span><span class="co">#&gt; imgSource():</span></span>
<span><span class="co">#&gt;   /__w/VoyagerHackathon/VoyagerHackathon/vignettes/vizgen_cellbound/images/mosaic_</span></span>
<span><span class="co">#&gt;   DAPI_z3.tif</span></span></code></pre></div>
<p>Use th <code>ext</code> function to get the extent of the image</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/ext.html" class="external-link">ext</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span>
<span><span class="co">#&gt;      xmin      xmax      ymin      ymax </span></span>
<span><span class="co">#&gt;  6499.909  6800.141 -1500.166 -1199.939</span></span></code></pre></div>
<p><code>SpatRasterImage</code> is a thin wrapper; use
<code>imgRaster</code> to get the image itself:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-class.html" class="external-link">imgRaster</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span>, col <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/pal_viridis.html" class="external-link">viridis_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">50</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-49-1.png" width="700"></p>
<p>The advantage of <code>SpatRasterImage</code> is that one can use
vector geometries such as in <code>sf</code> data frames to extract data
from the raster image. With a binary mask, the <code>terra</code>
package can convert the mask into polygons and vice versa.</p>
<p>But there are some disadvantages, such as that <code>terra</code> is
built for geography so it’s difficult to perform affine transforms of
the image (including rotation); in geography the transformation is
performed by reprojecting the map and there are standards for the
projections such as the Mercator and Robinson projections of the world
map. So when the <code>SpatRasterImage</code> is rotated, it’s converted
into <code>EBImage</code>, which can be converted back to
<code>SpatRasterImage</code>. <code>BioFormatsImage</code> can also be
converted into <code>SpatRasterImage</code> though that goes through
<code>EBImage</code>. Also, one cannot perform image processing such as
morphological operations, watershed segmentation, and so on that can be
performed by <code>EBImage</code>.</p>
<p>Here we convert this image into <code>EBImage</code>, and plot it
with the <code>EBImage</code> package</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ebi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/toEBImage.html" class="external-link">toEBImage</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-class.html" class="external-link">imgRaster</a></span><span class="op">(</span><span class="va">ebi</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/normalize.html" class="external-link">normalize</a></span><span class="op">(</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"raster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-50-1.png" width="700"></p>
<p>Finally, we can call a Voyager function to plot the image, cell
segmentation, and transcript spots together:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, features <span class="op">=</span> <span class="st">"volume"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   image_id <span class="op">=</span> <span class="st">"DAPI_z3"</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rg</span>, size <span class="op">=</span> <span class="fl">0.1</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<p>BTW, the reason why the transcript spots are not assigned to cells
but stored in <code>rowGeometries</code> as <code>MULTIPOINT</code> is
that some transcript spots are not in any segmented cells as shown in
this plot. However, it doesn’t mean that they are irrelevant.</p>
<p>Also, the plot above shows transcript spots from all genes in this
dataset and they’re very down sampled. In the hackathon, you may choose
to work on <a href="https://github.com/pachterlab/voyager/issues/14" class="external-link">this issue</a> to
add <code>rowGeometries</code> to the <code>plotSpatialFeature</code>
function and add an argument to specify which gene(s) to plot. Typically
only a small number of genes should be plotted at a time, because
otherwise the points will be too crowded to see and if using colors to
distinguish between genes, there will be colors that are difficult to
distinguish. I suggest using point shape to distinguish between genes
when colors are already used for cell data.</p>
<p>Turn the above code chunk into a function as we’ll call it several
times more though only for this MERFISH dataset</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_sfe_vizgen</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, features <span class="op">=</span> <span class="st">"volume"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   image_id <span class="op">=</span> <span class="st">"DAPI_z3"</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/rowGeometries.html" class="external-link">txSpots</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.1</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="spatial-operations">Spatial operations<a class="anchor" aria-label="anchor" href="#spatial-operations"></a>
</h2>
<div class="section level3">
<h3 id="bounding-box">Bounding box<a class="anchor" aria-label="anchor" href="#bounding-box"></a>
</h3>
<p>The bounding box of a geometry is the smallest rectangle that
contains this geometry, so you get minimum and maximum x coordinates and
y coordinates. We can find the bounding box of individual
<code>sf</code> data frames with <code>st_bbox</code> from the
<code>sf</code> package</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_bbox.html" class="external-link">st_bbox</a></span><span class="op">(</span><span class="va">rg</span><span class="op">)</span></span>
<span><span class="co">#&gt;      xmin      ymin      xmax      ymax </span></span>
<span><span class="co">#&gt;  6500.184 -1499.778  6799.719 -1200.041</span></span></code></pre></div>
<p>However, in an SFE object, there are multiple geometries, such as
cell centroids, cell segmentation, nucleus segmentation, tissue
boundary, transcript spots, and so on, and there are images. The
<code>bbox</code> function for SFE aggregates the bounding boxes of all
the geometries (and optionally images) to get an overall bounding box of
the SFE object:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_vizgen</span><span class="op">)</span></span>
<span><span class="co">#&gt;      xmin      ymin      xmax      ymax </span></span>
<span><span class="co">#&gt;  6498.553 -1502.179  6800.404 -1199.928</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># In this case the image is not larger than the geometries</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, include_image <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;      xmin      ymin      xmax      ymax </span></span>
<span><span class="co">#&gt;  6498.553 -1502.179  6800.404 -1199.928</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cropping">Cropping<a class="anchor" aria-label="anchor" href="#cropping"></a>
</h3>
<p>You can think of the SFE object as a stack of maps that are aligned,
like the <a href="https://apps.nationalmap.gov/viewer/" class="external-link">National Map</a>
layers of satellite images, land use, administrative boundaries,
watersheds, rock formations, faults, and etc. Cropping will crop all of
the maps. One can crop with either a bounding box or a polygon of any
shape. The <code>colGeometryName</code> argument specifies the
<code>colGeometry</code> to decide which cell to keep after cropping.
Using the centroid would be different from using the cell polygon since
a polygon can slightly overlap with the bounding box while the centroid
is outside.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">6550</span>, xmax <span class="op">=</span> <span class="fl">6650</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">1350</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">1250</span><span class="op">)</span></span>
<span><span class="va">sfe_cropped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/crop.html" class="external-link">crop</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, <span class="va">bbox_use</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_cropped</span><span class="op">)</span></span>
<span><span class="co">#&gt;  xmin  ymin  xmax  ymax </span></span>
<span><span class="co">#&gt;  6550 -1350  6650 -1250</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_sfe_vizgen</span><span class="op">(</span><span class="va">sfe_cropped</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-57-1.png" width="700"></p>
<p>Don’t like those tiny slivers of cells at the boundary of the
bounding box? We can also keep any cell that intersects with the
bounding box and with larger datasets, this is much faster than finding
the actual intersection geometries. The <code>keep_whole</code> argument
makes sure that the cells are kept whole; “col” because it can also be
“annot” to keep <code>annotGeometry</code> items (e.g. cell segmentation
in Visium datasets) whole.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_cropped2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/crop.html" class="external-link">crop</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, <span class="va">bbox_use</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                     keep_whole <span class="op">=</span> <span class="st">"col"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_sfe_vizgen</span><span class="op">(</span><span class="va">sfe_cropped2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sfc.html" class="external-link">st_as_sfc</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_bbox.html" class="external-link">st_bbox</a></span><span class="op">(</span><span class="va">bbox_use</span><span class="op">)</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>            linewidth <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-59-1.png" width="700"></p>
<p>Here the original bounding box is shown in red and the cells that
partially overlap are not cropped.</p>
<p>We can also only keep cells covered by (i.e. entirely within) the
bounding box</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_cropped3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/crop.html" class="external-link">crop</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, <span class="va">bbox_use</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                     keep_whole <span class="op">=</span> <span class="st">"col"</span>, cover <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_sfe_vizgen</span><span class="op">(</span><span class="va">sfe_cropped3</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-61-1.png" width="700"></p>
<p>We can also use a geometry to remove a part of the data by specifying
<code>op = st_difference</code>; <code>keep_whole</code> and
<code>cover</code> still apply.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_hole</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/crop.html" class="external-link">crop</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, <span class="va">bbox_use</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                 op <span class="op">=</span> <span class="va">st_difference</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_sfe_vizgen</span><span class="op">(</span><span class="va">sfe_hole</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-63-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="transformation">Transformation<a class="anchor" aria-label="anchor" href="#transformation"></a>
</h3>
<p>We can rotate (right now only multiples of 90 degrees), mirror,
transpose, and translate the SFE object, such as when there’s a
canonical orientation like in brain sections but the data is of a
different orientation when read in. Here all geometries and images are
transformed while keeping them aligned.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_mirror</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SFE-transform.html" class="external-link">mirror</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_sfe_vizgen</span><span class="op">(</span><span class="va">sfe_mirror</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-65-1.png" width="700"></p>
<p>Individual images can be transformed say in case it doesn’t initially
align with the geometries, though for this dataset, this will put the
image out of alignment</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-class.html" class="external-link">mirrorImg</a></span><span class="op">(</span><span class="va">sfe_vizgen</span>, image_id <span class="op">=</span> <span class="st">"DAPI_z3"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_sfe_vizgen</span><span class="op">(</span><span class="va">sfe_img</span><span class="op">)</span></span></code></pre></div>
<p><img src="sfe_files/figure-html/unnamed-chunk-67-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="multiple-samples">Multiple samples<a class="anchor" aria-label="anchor" href="#multiple-samples"></a>
</h2>
<p>Thus far, the example dataset used only has one sample. The
<code>SpatialExperiment</code> (SPE) object has a special column in
<code>colData</code> called <code>sample_id</code>, so data from
multiple tissue sections can coexist in the same SPE object for joint
dimension reduction and clustering while keeping the spatial coordinates
separate. It’s important to keep spatial coordinates of different tissue
sections separate because first, the coordinates would only make sense
within the same section, and second, the coordinates from different
sections can have overlapping numeric values.</p>
<p>SFE inherits from SPE, and with geometries and spatial graphs,
<code>sample_id</code> is even more important. The geometry and graph
getter and setter functions have a <code>sample_id</code> argument,
which is optional when only one sample is present in the SFE object.
This argument is mandatory if multiple samples are present, and can be a
character vector for multiple samples or “all” for all samples. Below
are examples of using the getters and setters for multiple samples.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Construct toy dataset with 2 samples</span></span>
<span><span class="va">sfe1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small"</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="va">sfe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe2</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="st">"sample02"</span></span>
<span><span class="op">(</span><span class="va">sfe_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sfe1</span>, <span class="va">sfe2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 149 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(149): AAATTACCTATCGATG AACATATCAACTGGTG ... TTCCTCGGACTAACCA</span></span>
<span><span class="co">#&gt;   TTCTGACCGGGCTCAA</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (GEOMETRY), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A: </span></span>
<span><span class="co">#&gt; sample02:</span></span></code></pre></div>
<p>Use the <code>sampleIDs</code> function to see the names of all
samples</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/sampleIDs.html" class="external-link">sampleIDs</a></span><span class="op">(</span><span class="va">sfe_combined</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Vis5A"    "sample02"</span></span></code></pre></div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only get the geometries for the second sample</span></span>
<span><span class="op">(</span><span class="va">spots2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"spotPoly"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 72 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 6000 ymin: 7025.865 xmax: 8000 ymax: 9000</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                  sample_id                       geometry</span></span>
<span><span class="co">#&gt; AACACACGCTCGCCGC  sample02 POLYGON ((6597.869 7842.575...</span></span>
<span><span class="co">#&gt; AACCGCTAAGGGATGC  sample02 POLYGON ((6724.811 9000, 67...</span></span>
<span><span class="co">#&gt; AACGCTGTTGCTGAAA  sample02 POLYGON ((6457.635 7118.991...</span></span>
<span><span class="co">#&gt; AACGGACGTACGTATA  sample02 POLYGON ((6737.064 8083.571...</span></span>
<span><span class="co">#&gt; AATAGAATCTGTTTCA  sample02 POLYGON ((7570.153 8564.368...</span></span>
<span><span class="co">#&gt; ACAAATCGCACCGAAT  sample02 POLYGON ((8000 7997.001, 79...</span></span>
<span><span class="co">#&gt; ACAATTGTGTCTCTTT  sample02 POLYGON ((6043.169 7843.77,...</span></span>
<span><span class="co">#&gt; ACAGGCTTGCCCGACT  sample02 POLYGON ((7428.88 7358.195,...</span></span>
<span><span class="co">#&gt; ACCAGTGCGGGAGACG  sample02 POLYGON ((6460.753 8566.757...</span></span>
<span><span class="co">#&gt; ACCCTCCCTTGCTATT  sample02 POLYGON ((7847.503 8563.771...</span></span></code></pre></div>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only set the geometries for the second sample</span></span>
<span><span class="co"># Leaving geometries of the first sample intact</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"spotPoly"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots2</span></span></code></pre></div>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph only for the second sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"foo"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get graph only for the second sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"foo"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 72 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 406 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.83179 </span></span>
<span><span class="co">#&gt; Average number of links: 5.638889 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 72 5184 72 25.82104 289.8299</span></span></code></pre></div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph of the same name for both samples</span></span>
<span><span class="co"># The graphs are computed separately for each sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraphs</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span>, name <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get multiple graphs of the same name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraphs</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span>, name <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $Vis5A</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 394 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 6.645303 </span></span>
<span><span class="co">#&gt; Average number of links: 5.116883 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 31.68056 311.7544</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample02</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 72 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 366 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.060185 </span></span>
<span><span class="co">#&gt; Average number of links: 5.083333 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 72 5184 72 29.83889 291.5833</span></span></code></pre></div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Or just all graphs of the margin</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraphs</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $Vis5A</span></span>
<span><span class="co">#&gt; $Vis5A$visium</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 394 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 6.645303 </span></span>
<span><span class="co">#&gt; Average number of links: 5.116883 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 31.68056 311.7544</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample02</span></span>
<span><span class="co">#&gt; $sample02$foo</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 72 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 406 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.83179 </span></span>
<span><span class="co">#&gt; Average number of links: 5.638889 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 72 5184 72 25.82104 289.8299</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sample02$visium</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 72 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 366 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.060185 </span></span>
<span><span class="co">#&gt; Average number of links: 5.083333 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 72 5184 72 29.83889 291.5833</span></span></code></pre></div>
<p>Sample IDs can also be changed, with the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/changeSampleIDs.html" class="external-link">changeSampleIDs()</a></code> function, with a named vector whose names
are the old names and values are the new names.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/changeSampleIDs.html" class="external-link">changeSampleIDs</a></span><span class="op">(</span><span class="va">sfe_combined</span>, </span>
<span>                                replacement <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Vis5A <span class="op">=</span> <span class="st">"foo"</span>, sample02 <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sfe_combined</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 149 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(149): AAATTACCTATCGATG AACATATCAACTGGTG ... TTCCTCGGACTAACCA</span></span>
<span><span class="co">#&gt;   TTCTGACCGGGCTCAA</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (GEOMETRY), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; foo: col: visium</span></span>
<span><span class="co">#&gt; bar: col: foo, visium</span></span></code></pre></div>
<p>Basically, all the functions covered above have an argument
<code>sample_id</code> if the operation is not to be performed on all
samples. Or set <code>sample_id = "all"</code> to perform on all
samples.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/sampleIDs.html" class="external-link">sampleIDs</a></span><span class="op">(</span><span class="va">sfe_combined</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "foo" "bar"</span></span></code></pre></div>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span>
<span><span class="co">#&gt;        foo  bar</span></span>
<span><span class="co">#&gt; xmin  5000 6000</span></span>
<span><span class="co">#&gt; ymin 13000 7000</span></span>
<span><span class="co">#&gt; xmax  7000 8000</span></span>
<span><span class="co">#&gt; ymax 15000 9000</span></span></code></pre></div>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_combined</span>, sample_id <span class="op">=</span> <span class="st">"foo"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  xmin  ymin  xmax  ymax </span></span>
<span><span class="co">#&gt;  5000 13000  7000 15000</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="future-directions">Future directions<a class="anchor" aria-label="anchor" href="#future-directions"></a>
</h2>
<p>See the <a href="https://github.com/pachterlab/SpatialFeatureExperiment/issues" class="external-link">GitHub
issues</a></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2024-02-28 r85999)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] scales_1.3.0                   EBImage_4.45.0                </span></span>
<span><span class="co">#&gt;  [3] Voyager_1.5.0                  SFEData_1.5.0                 </span></span>
<span><span class="co">#&gt;  [5] SpatialFeatureExperiment_1.5.2 scran_1.31.0                  </span></span>
<span><span class="co">#&gt;  [7] scater_1.31.2                  scuttle_1.13.0                </span></span>
<span><span class="co">#&gt;  [9] SingleCellExperiment_1.25.0    SummarizedExperiment_1.33.3   </span></span>
<span><span class="co">#&gt; [11] Biobase_2.63.0                 GenomicRanges_1.55.3          </span></span>
<span><span class="co">#&gt; [13] GenomeInfoDb_1.39.6            IRanges_2.37.1                </span></span>
<span><span class="co">#&gt; [15] S4Vectors_0.41.3               BiocGenerics_0.49.1           </span></span>
<span><span class="co">#&gt; [17] MatrixGenerics_1.15.0          matrixStats_1.2.0             </span></span>
<span><span class="co">#&gt; [19] ggplot2_3.5.0                  terra_1.7-71                  </span></span>
<span><span class="co">#&gt; [21] sf_1.0-15                      BiocStyle_2.31.0              </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7              filelock_1.0.3           </span></span>
<span><span class="co">#&gt;   [3] tibble_3.2.1              R.oo_1.26.0              </span></span>
<span><span class="co">#&gt;   [5] lifecycle_1.0.4           edgeR_4.1.17             </span></span>
<span><span class="co">#&gt;   [7] lattice_0.22-5            magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [9] limma_3.59.3              sass_0.4.8               </span></span>
<span><span class="co">#&gt;  [11] rmarkdown_2.25            jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [13] yaml_2.3.8                metapod_1.11.1           </span></span>
<span><span class="co">#&gt;  [15] sp_2.1-3                  DBI_1.2.2                </span></span>
<span><span class="co">#&gt;  [17] abind_1.4-5               zlibbioc_1.49.0          </span></span>
<span><span class="co">#&gt;  [19] purrr_1.0.2               R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [21] RCurl_1.98-1.14           rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt;  [23] GenomeInfoDbData_1.2.11   ggrepel_0.9.5            </span></span>
<span><span class="co">#&gt;  [25] irlba_2.3.5.1             units_0.8-5              </span></span>
<span><span class="co">#&gt;  [27] RSpectra_0.16-1           dqrng_0.3.2              </span></span>
<span><span class="co">#&gt;  [29] pkgdown_2.0.7             DelayedMatrixStats_1.25.1</span></span>
<span><span class="co">#&gt;  [31] codetools_0.2-19          DropletUtils_1.23.0      </span></span>
<span><span class="co">#&gt;  [33] DelayedArray_0.29.8       tidyselect_1.2.0         </span></span>
<span><span class="co">#&gt;  [35] farver_2.1.1              ScaledMatrix_1.11.0      </span></span>
<span><span class="co">#&gt;  [37] viridis_0.6.5             BiocFileCache_2.11.1     </span></span>
<span><span class="co">#&gt;  [39] jsonlite_1.8.8            BiocNeighbors_1.21.2     </span></span>
<span><span class="co">#&gt;  [41] e1071_1.7-14              systemfonts_1.0.5        </span></span>
<span><span class="co">#&gt;  [43] ggnewscale_0.4.10         tools_4.4.0              </span></span>
<span><span class="co">#&gt;  [45] ragg_1.2.7                sfarrow_0.4.1            </span></span>
<span><span class="co">#&gt;  [47] Rcpp_1.0.12               glue_1.7.0               </span></span>
<span><span class="co">#&gt;  [49] gridExtra_2.3             SparseArray_1.3.4        </span></span>
<span><span class="co">#&gt;  [51] xfun_0.42                 dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [53] HDF5Array_1.31.6          withr_3.0.0              </span></span>
<span><span class="co">#&gt;  [55] BiocManager_1.30.22       fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [57] boot_1.3-30               rhdf5filters_1.15.2      </span></span>
<span><span class="co">#&gt;  [59] bluster_1.13.0            fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [61] spData_2.3.0              digest_0.6.34            </span></span>
<span><span class="co">#&gt;  [63] rsvd_1.0.5                mime_0.12                </span></span>
<span><span class="co">#&gt;  [65] R6_2.5.1                  textshaping_0.3.7        </span></span>
<span><span class="co">#&gt;  [67] colorspace_2.1-0          wk_0.9.1                 </span></span>
<span><span class="co">#&gt;  [69] jpeg_0.1-10               RSQLite_2.3.5            </span></span>
<span><span class="co">#&gt;  [71] R.methodsS3_1.8.2         utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [73] generics_0.1.3            data.table_1.15.2        </span></span>
<span><span class="co">#&gt;  [75] class_7.3-22              httr_1.4.7               </span></span>
<span><span class="co">#&gt;  [77] htmlwidgets_1.6.4         S4Arrays_1.3.4           </span></span>
<span><span class="co">#&gt;  [79] spdep_1.3-3               pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [81] scico_1.5.0               gtable_0.3.4             </span></span>
<span><span class="co">#&gt;  [83] blob_1.2.4                XVector_0.43.1           </span></span>
<span><span class="co">#&gt;  [85] htmltools_0.5.7           bookdown_0.38            </span></span>
<span><span class="co">#&gt;  [87] fftwtools_0.9-11          png_0.1-8                </span></span>
<span><span class="co">#&gt;  [89] SpatialExperiment_1.13.0  knitr_1.45               </span></span>
<span><span class="co">#&gt;  [91] rjson_0.2.21              curl_5.2.1               </span></span>
<span><span class="co">#&gt;  [93] proxy_0.4-27              cachem_1.0.8             </span></span>
<span><span class="co">#&gt;  [95] rhdf5_2.47.4              stringr_1.5.1            </span></span>
<span><span class="co">#&gt;  [97] BiocVersion_3.19.1        KernSmooth_2.23-22       </span></span>
<span><span class="co">#&gt;  [99] parallel_4.4.0            vipor_0.4.7              </span></span>
<span><span class="co">#&gt; [101] arrow_14.0.2.1            AnnotationDbi_1.65.2     </span></span>
<span><span class="co">#&gt; [103] desc_1.4.3                s2_1.1.6                 </span></span>
<span><span class="co">#&gt; [105] pillar_1.9.0              grid_4.4.0               </span></span>
<span><span class="co">#&gt; [107] vctrs_0.6.5               BiocSingular_1.19.0      </span></span>
<span><span class="co">#&gt; [109] dbplyr_2.4.0              beachmat_2.19.1          </span></span>
<span><span class="co">#&gt; [111] sfheaders_0.4.4           cluster_2.1.6            </span></span>
<span><span class="co">#&gt; [113] beeswarm_0.4.0            evaluate_0.23            </span></span>
<span><span class="co">#&gt; [115] magick_2.8.3              cli_3.6.2                </span></span>
<span><span class="co">#&gt; [117] locfit_1.5-9.9            compiler_4.4.0           </span></span>
<span><span class="co">#&gt; [119] rlang_1.1.3               crayon_1.5.2             </span></span>
<span><span class="co">#&gt; [121] labeling_0.4.3            classInt_0.4-10          </span></span>
<span><span class="co">#&gt; [123] fs_1.6.3                  ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt; [125] stringi_1.8.3             viridisLite_0.4.2        </span></span>
<span><span class="co">#&gt; [127] deldir_2.0-4              BiocParallel_1.37.0      </span></span>
<span><span class="co">#&gt; [129] assertthat_0.2.1          Biostrings_2.71.2        </span></span>
<span><span class="co">#&gt; [131] munsell_0.5.0             tiff_0.1-12              </span></span>
<span><span class="co">#&gt; [133] Matrix_1.6-5              ExperimentHub_2.11.1     </span></span>
<span><span class="co">#&gt; [135] patchwork_1.2.0           sparseMatrixStats_1.15.0 </span></span>
<span><span class="co">#&gt; [137] bit64_4.0.5               Rhdf5lib_1.25.1          </span></span>
<span><span class="co">#&gt; [139] KEGGREST_1.43.0           statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [141] highr_0.10                AnnotationHub_3.11.1     </span></span>
<span><span class="co">#&gt; [143] igraph_2.0.2              memoise_2.0.1            </span></span>
<span><span class="co">#&gt; [145] bslib_0.6.1               bit_4.0.5</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Alik Huseynov, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
