<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="VoyagerHackathon">
<title>Exploratory spatial data analysis with Voyager • VoyagerHackathon</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Exploratory spatial data analysis with Voyager">
<meta property="og:description" content="VoyagerHackathon">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">VoyagerHackathon</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/info.html">Event info</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-workshop">Workshop</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-workshop">
    <a class="dropdown-item" href="../articles/sfe.html">SFE</a>
    <a class="dropdown-item" href="../articles/voyager.html">Voyager</a>
    <a class="dropdown-item" href="../articles/package.html">R package dev</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/lambdamoses/VoyagerHackathon/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Exploratory spatial data analysis with Voyager</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:dlu2@caltech.edu" class="email">dlu2@caltech.edu</a>
      
                              <h4 data-toc-skip class="author">Lior
Pachter</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:lpachter@caltech.edu" class="email">lpachter@caltech.edu</a>
      
                  
            <h4 data-toc-skip class="date">2024-03-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lambdamoses/VoyagerHackathon/blob/HEAD/vignettes/voyager.Rmd" class="external-link"><code>vignettes/voyager.Rmd</code></a></small>
      <div class="d-none name"><code>voyager.Rmd</code></div>
    </div>

    
    
<p>The following R packages are used in this workshop, which are all on
CRAN or Bioconductor. Bioconductor 3.19 (devel) is used in the
workshop.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexcb/rjson" class="external-link">rjson</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thomasp85/scico" class="external-link">scico</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/BiocSingular" class="external-link">BiocSingular</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="part-2-voyager-esda">Part 2: Voyager ESDA<a class="anchor" aria-label="anchor" href="#part-2-voyager-esda"></a>
</h2>
<p><img src="voyager.png" alt="Schematics of the Voyager project" width="100%"></p>
<div class="section level3">
<h3 id="x-visium">10X Visium<a class="anchor" aria-label="anchor" href="#x-visium"></a>
</h3>
<p>As Visium from 10X Genomics is currently the most popular spatial
transcriptomics technology, this workshop uses a Visium dataset.</p>
<p><img src="https://pachterlab.github.io/LP_2021/04-current_files/figure-html/n-insts-1.png" alt='Bar chart showing the number of institutions using each spatial transcriptomics data collection method. Only methods used by at least 3 different institutions are shown. The bars are colored by category of the methods. 10X Visium which is based on sequence barcoding is used by almost 200 institutions. Following Visium are GeoMX DSP and GeoMX WTA which are "regions of interest" (ROI) methods, along with 2016 ST, MERFISH, and some other ROI selection methods.'></p>
<p>In Visium, capture sequences with spot barcode, unique molecule
identifier (UMI), and polyT to capture poly-adenylated mRNAs are printed
in a hexagonal array on a glass slide. Each spot barcode has a known
location, and the spots are 55 <span class="math inline">\(\mu
m\)</span> in diameter and 100 <span class="math inline">\(\mu
m\)</span> apart center to center. As the spots are much larger than
most types of cells, Visium does not have single cell resolution. Tissue
is mounted on each of the 4 capture areas on the slide, and each capture
area has 4992 spots. The spots capture the transcripts from the tissue,
which are then reverse transcribed, amplified, and sequenced.</p>
<p><img src="visium.png" alt="Intro to Visium, described in the previous paragraph" width="833"></p>
<p>Space Ranger is the official software to align the sequencing reads
to the genome and quantify the UMIs in each spot for each gene. Spatial
Ranger also takes in a histology image of the capture area, with which
it determines which spots are in tissue. The user can also manually
determine which spots are in tissue in the Loupe Browser.</p>
</div>
<div class="section level3">
<h3 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h3>
<p>The dataset used in this vignette is from the paper <a href="https://doi.org/10.1038/s42003-021-02810-x" class="external-link">Large-scale
integration of single-cell transcriptomic data captures transitional
progenitor states in mouse skeletal muscle regeneration</a> <span class="citation">(McKellar et al. 2021)</span>. Notexin was injected
into the tibialis anterior muscle of mice to induce injury, and the
healing muscle was collected 2, 5, and 7 days post injury for Visium
analysis. The dataset in this vignette is from the timepoint at day 2.
The vignette starts with a <code>SpatialFeatureExperiment</code> (SFE)
object.</p>
<p>The gene count matrix was directly downloaded <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4904759" class="external-link">from
GEO</a>. All 4992 spots, whether in tissue or not, are included. The
H&amp;E image was used for nuclei and myofiber segmentation. A subset of
nuclei from randomly selected regions from all 3 timepoints were
manually annotated to train a StarDist model to segment the rest of the
nuclei, and the myofibers were all manually segmented. The tissue
boundary was found by thresholding in OpenCV, and small polygons were
removed as they are likely to be debris. Spot polygons were constructed
with the spot centroid coordinates and diameter in the Space Ranger
output. The <code>in_tissue</code> column in <code>colData</code>
indicates which spot polygons intersect the tissue polygons, and is
based on <code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>.</p>
<p>Tissue boundary, nuclei, myofiber, and Visium spot polygons are
stored as <code>sf</code> data frames in the SFE object. See <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SpatialFeatureExperiment/inst/doc/SFE.html" class="external-link">the
vignette of <code>SpatialFeatureExperiment</code></a> for more details
on the structure of the SFE object. The SFE object of this dataset is
provided in the <code>SFEData</code> package; we begin by downloading
the data and loading it into R.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC AAACAAGTATCTCCCA ... TTGTTTGTATTACACG</span></span>
<span><span class="co">#&gt;   TTGTTTGTGTAAATTC</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (POLYGON), myofiber_simplified (POLYGON), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<p>The H&amp;E image of this section:
<img src="tissue_lowres_5a.jpeg" alt="A cross section of mouse muscle is slightly off center to the lower left. In the middle of the tissue is the notexin injury site with leukocyte infiltration and fewer myofibers. The rest of the tissue section is tightly packed with myofibers." width="512"></p>
<p>The image can be added to the SFE object and plotted behind the
geometries, and needs to be flipped to align to the spots because the
origin is at the top left for the image but bottom left for
geometries.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">addImg</span><span class="op">(</span><span class="va">sfe</span>, imageSource <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span>, </span>
<span>              image_id <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>              scale_fct <span class="op">=</span> <span class="fl">1024</span><span class="op">/</span><span class="fl">22208</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-class.html" class="external-link">mirrorImg</a></span><span class="op">(</span><span class="va">sfe</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="exploratory-data-analysis">Exploratory data analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h3>
<div class="section level4">
<h4 id="spots-in-tissue">Spots in tissue<a class="anchor" aria-label="anchor" href="#spots-in-tissue"></a>
</h4>
<p>While the example dataset has all Visium spots whether on tissue or
not, only spots that intersect tissue are used for further analyses.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode"   "col"       "row"       "x"         "y"         "dia"      </span></span>
<span><span class="co">#&gt;  [7] "tissue"    "sample_id" "nCounts"   "nGenes"    "prop_mito" "in_tissue"</span></span></code></pre></div>
<p>Total UMI counts (<code>nCounts</code>), number of genes detected per
spot (<code>nGenes</code>), and the proportion of mitochondrially
encoded counts (<code>prop_mito</code>) have been precomputed and are in
<code>colData(sfe)</code>. The <code><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature()</a></code>
function can be used to visualize various attributes in space: the
expression of any gene, <code>colData</code> values, and geometry
attributes in <code>colGeometry</code> and <code>annotGeometry</code>.
The Visium spots are plotted as polygons reflecting their actual size
relative to the tissue, rather than as points, as is the case in other
packages that plot Visium data. The plotting of geometries is being
performed under the hood with <code>geom_sf</code>.</p>
<p>The tissue boundary was found by thresholding the H&amp;E image and
removing small polygons that are most likely debris. The
<code>in_tissue</code> column of <code>colData(sfe)</code> indicates
which Visium spot polygon intersects the tissue polygon; this can be
found with <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">SpatialFeatureExperiment::annotPred()</a></code>.</p>
<p>We demonstrate the use of <code>scran</code> <span class="citation">(Lun, Bach, and Marioni 2016)</span> for normalization
below, although we note that it is not necessarily the best approach to
normalizing spatial transcriptomics data. The problem of when and how to
normalize spatial transcriptomics data is non-trivial because, as the
<code>nCounts</code> plot in space shows above, spatial autocorrelation
is evident. Furthermore, in Visium, reverse transcription occurs in situ
on the spots, but PCR amplification occurs after the cDNA is dissociated
from the spots. Artifacts may be subsequently introduced from the
amplification step, and these would not be associated with spatial
origin. Spatial artifacts may arise from the diffusion of transcripts
and tissue permeablization. However, given how the total counts seem to
correspond to histological regions, the total counts may have a
biological component and hence should not be treated as a technical
artifact to be normalized away as in scRNA-seq data normalization
methods. In other words, the issue of normalization for spatial
transcriptomics data, and Visium in particular, is complex and is
currently unsolved.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#clusters &lt;- quickCluster(sfe_tissue)</span></span>
<span><span class="co">#sfe_tissue &lt;- computeSumFactors(sfe_tissue, clusters=clusters)</span></span>
<span><span class="co">#sfe_tissue &lt;- sfe_tissue[, sizeFactors(sfe_tissue) &gt; 0]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p>Myofiber and nuclei segmentation polygons are available in this
dataset in the <code>annotGeometries</code> field. Myofibers were
manually segmented, and nuclei were segmented with <a href="https://github.com/stardist/stardist" class="external-link"><code>StarDist</code></a>
trained with a manually segmented subset.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "tissueBoundary"      "myofiber_full"       "myofiber_simplified"</span></span>
<span><span class="co">#&gt; [4] "nuclei"              "nuclei_centroid"</span></span></code></pre></div>
<div class="section level5">
<h5 id="from-myofibers-and-nuclei-to-visium-spots">From myofibers and nuclei to Visium spots<a class="anchor" aria-label="anchor" href="#from-myofibers-and-nuclei-to-visium-spots"></a>
</h5>
<p>The <code><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature()</a></code> function can also be used to
plot attributes of geometries, i.e. the non-geometry columns in the
<code>sf</code> data frames in the <code>rowGeometries</code>,
<code>colGeometries</code>, or <code>annotGeometries</code> fields of
the SFE object.</p>
<p>The myofiber polygons from <code>annotGeometries</code> can be
plotted as shown below, colored by cross section area as observed in the
tissue section. The <code>aes_use</code> argument is set to
<code>color</code> rather than <code>fill</code> (default for polygons)
to only plot the Visium spot outlines to make the myofiber polygons more
visible. The <code>fill</code> argument is set to <code>NA</code> to
make the Visium spots look hollow, and the <code>size</code> argument
controls the thickness of the outlines. The <code>annot_aes</code>
argument specifies which column in the <code>annotGeometry</code> to use
to specify the values of an aesthstic, just like <code>aes</code> in
<code>ggplot2</code> (<code>aes_string</code> to be precise, since
<code>tidyeval</code> is not used here). The <code>annot_fixed</code>
argument (not used here) can set the fixed size, alpha, color, and etc.
for the <code>annotGeometry</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"nCounts"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span>, fill <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-13-1.png" alt="Plot of Visium spots in tissue and myofiber polygons in physical space. Visium spots are colored by nCounts, and myofibers are colored by area." width="700"></p>
<p>The larger myofibers seem to have fewer total counts, possibly
because the larger size of these myofibers dilutes the transcripts. This
hints at the need for a normalization procedure.</p>
<p>With <code>SpatialFeatureExperiment</code>, we can find the number of
myofibers and nuclei that intersect each Visium spot. The predicate can
be <a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">anything
implemented in <code>sf</code></a>, so for example, the number of nuclei
fully covered by each Visium spot can also be found. The default
predicate is <code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">n_myofibers</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">annotNPred</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>             annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"n_myofibers"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, image <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                   linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-15-1.png" alt="Plot of Visium spots in tissue in physical space, colored by number of myofibers intersecting each spot." width="700"></p>
<p>There is no one-to-one mapping between Visium spots and myofibers.
However, we can relate attributes of myofibers to gene expression
detected at the Visium spots. One way to do so is to summarize the
attributes of all myofibers that intersect (or choose another better
predicate implemented in <code>sf</code>) each spot, such as to
calculate the mean, median, or sum. This can be done with the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotSummary.html" class="external-link">annotSummary()</a></code> function in
<code>SpatialFeatureExperiment</code>. The default predicate is
<code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>, and the default summary function is
<code><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">mean_myofiber_area</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotSummary.html" class="external-link">annotSummary</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"spotPoly"</span>, <span class="st">"myofiber_simplified"</span>, </span>
<span>               annotColNames <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="co"># it always returns a data frame</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The gray spots don't intersect any myofiber</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"mean_myofiber_area"</span>, <span class="st">"spotPoly"</span>, image <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                   color <span class="op">=</span> <span class="st">"black"</span>, linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-17-1.png" alt="Plot of Visium spots in tissue in physical space, colored by the average area of myofibers that intersect each spot. The average area is higher near the mid-top right part of the tissue." width="700"></p>
<p>This reveals the relationship between the mean area of myofibers
intersecting each Visium spot and other aspects of the spots, such as
total counts and gene expression.</p>
<p>The NAs (gray) designate spots not intersecting any myofibers,
e.g. those in the inflammatory region.</p>
<p>The nGenes vs. nCounts plot is a standard QC plot in scRNA-seq, but
here we see two mysterious branches and two clusters in the nGenes
vs. nCounts plot and the proportion of mitochondrial counts vs. nCounts
plot. The two branches or clusters seem to be related to myofiber
size.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span>, colour_by <span class="op">=</span> <span class="st">"mean_myofiber_area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, colour_by <span class="op">=</span> <span class="st">"mean_myofiber_area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="section level6">
<h6 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h6>
<ol style="list-style-type: decimal">
<li>Use the <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">annotNPred()</a></code> function to find the number of
nuclei intersecting each Visium spot. The nuclei segmentation polygons
are in the <code>annotGeometry</code> called “nuclei”.</li>
<li>Color the Visium spots with the number of nuclei. Which histological
region tends to have more nuclei per Visium spot?</li>
<li>Does the number of nuclei per spot correlate with nCounts?</li>
</ol>
</div>
</div>
<div class="section level5">
<h5 id="plot-gene-expression-in-space">Plot gene expression in space<a class="anchor" aria-label="anchor" href="#plot-gene-expression-in-space"></a>
</h5>
<p>Marker genes: Myh7 (Type I, slow twitch, aerobic), Myh2 (Type IIa,
fast twitch, somewhat aerobic), Myh4 (Type IIb, fast twitch, anareobic),
Myh1 (Type IIx, fast twitch, anaerobic), from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5526362/" class="external-link">this
protocol</a> <span class="citation">(Wang, Yue, and Kuang
2017)</span></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>I <span class="op">=</span> <span class="st">"Myh7"</span>, IIa <span class="op">=</span> <span class="st">"Myh2"</span>, IIb <span class="op">=</span> <span class="st">"Myh4"</span>, IIx <span class="op">=</span> <span class="st">"Myh1"</span><span class="op">)</span></span></code></pre></div>
<p>We first examine the Type I myofibers. This is a fast twitch muscle,
so we don’t expect many slow twitch Type I myofibers. Row names in
<code>sfe_tissue</code> are Ensembl IDs in order to avoid ambiguity as
sometimes multiple Ensembl IDs have the same gene symbol and some genes
have aliases. However, gene symbols are shorter and more human readable
than Ensembl IDs, and are better suited to display on plots. In the
<code><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature()</a></code> function and other functions in
<code>Voyager</code>, even when the row names are recorded as Ensembl
IDs, the <code>features</code> argument can take gene symbols if when a
column in <code>rowData(sfe)</code> that has the gene symbols are
supplied in the <code>swap_rownames</code> argument. All function in
<code>Voyager</code> that queries genes has the
<code>swap_rownames</code> argument.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh2"</span>, <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, </span>
<span>                   exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="section level6">
<h6 id="exercises-1">Exercises<a class="anchor" aria-label="anchor" href="#exercises-1"></a>
</h6>
<ol style="list-style-type: decimal">
<li>The <code>exprs_values</code> argument specifies the assay to use,
which is by default “logcounts”, i.e. the log normalized data. This
default may or may not be suitable in practice given that total UMI
counts may have biological relevance in spatial data. Plot one of the
marker genes above, but with the “counts” assay.</li>
<li>Look up the documentation of <code><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature()</a></code>. Try
plotting the Visium spots as filled circles that are partially
transparent.</li>
</ol>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="spatial-neighborhood-graphs">Spatial neighborhood graphs<a class="anchor" aria-label="anchor" href="#spatial-neighborhood-graphs"></a>
</h3>
<p>A spatial neighborhood graph is required to compute spatial
dependency metrics such as Moran’s I and Geary’s C. The
<code>SpatialFeatureExperiment</code> package wraps methods in
<code>spdep</code> to find spatial neighborhood graphs, which are stored
within the SFE object (see <code>spdep</code> documentation for
<code>gabrielneigh()</code>, <code>knearneigh()</code>,
<code>poly2nb()</code>, and <code>tri2nb()</code>). The
<code>Voyager</code> package then uses these graphs for spatial
dependency analyses, again based on <code>spdep</code> in this first
version, but methods from other geospatial packages, some of which also
use the spatial neighborhood graphs, may be added later.</p>
<p>For Visium, where the spots are in a hexagonal grid, the spatial
neighborhood graph is straightforward. However, for spatial technologies
with single cell resolution, e.g. MERFISH, different methods can be used
to find the spatial neighborhood graph. In this example, the method
“poly2nb” was used for myofibers, and it identifies myofiber polygons
that physically touch each other. <code>zero.policy = TRUE</code> will
allow for singletons, i.e. nodes without neighbors in the graph; in the
inflamed region, there are more singletons. We have not yet benchmarked
spatial neighborhood construction methods to determine which is the
“best” for different technologies; the particular method used here is
for demonstration purposes and may not be the best in practice:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">annotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_poly2nb"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, MARGIN <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       method <span class="op">=</span> <span class="st">"poly2nb"</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph()</a></code> function plots the graph in space
associated with a <code>colGeometry</code>, along with the geometry of
interest.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-23-1.png" alt="Spatial neighborhood graph of Visium spots that intersect tissue." width="700"></p>
<p>Similarly, the <code><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotAnnotGraph()</a></code> function plots the graph
associated with an <code>annotGeometry</code>, along with the geometry
of interest.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotAnnotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>               annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-24-1.png" alt="Spatial neighborhood graph of myofibers, where each edge connects two myofibers that touch." width="700"></p>
<p>There is no <code>plotRowGraph</code> yet since we haven’t worked
with a dataset where spatial graphs related to genes are relevant,
although the SFE object supports row graphs.</p>
</div>
<div class="section level3">
<h3 id="exploratory-spatial-data-analysis">Exploratory <em>spatial</em> data analysis<a class="anchor" aria-label="anchor" href="#exploratory-spatial-data-analysis"></a>
</h3>
<p>All spatial autocorrelation metrics in this package can be computed
directly on a vector or a matrix rather than an SFE object. The user
interface emulates those of dimension reductions in the
<code>scater</code> package (e.g. <code><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">calculateUMAP()</a></code> that
takes in a matrix or SCE object and returns a matrix, and
<code><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP()</a></code> that takes in an SCE object and adds the results
to the <code>reducedDims</code> field of the SCE object). So
<code>calculate*</code> functions take in a matrix or an SFE object and
directly return the results (format of the results depends on the
structure of the results), while <code>run*</code> functions take in an
SFE object and add the results to the object. In addition,
<code>colData*</code> functions compute the metrics for numeric
variables in <code>colData</code>. <code>colGeometry*</code> functions
compute the metrics for numeric columns in a <code>colGeometry</code>.
<code>annotGeometry*</code> functions compute the metrics for numeric
columns in a <code>annotGeometry</code>.</p>
</div>
<div class="section level3">
<h3 id="univariate-global">Univariate global<a class="anchor" aria-label="anchor" href="#univariate-global"></a>
</h3>
<p><code>Voyager</code> supports many univariate global spatial
autocorrelation implemented in <code>spdep</code> for ESDA: Moran’s I
and Geary’s C, permutation testing for Moran’s I and Geary’s C, Moran
plot, and correlograms. In addition, beyond <code>spdep</code>,
<code>Voyager</code> can cluster Moran plots and correlograms. Plotting
functions taking in SFE objects are implemented to plot the results with
<code>ggplot2</code> and with more customization options than
<code>spdep</code> plotting functions. The functions
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">calculateUnivariate()</a></code> (can take data outside SFE objects),
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate()</a></code> (for gene expression),
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colDataUnivariate()</a></code>, <code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colGeometryUnivariate()</a></code>,
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">annotGeometryUnivariate()</a></code>, and
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">reducedDimUnivariate()</a></code> compute univariate spatial
statistics for different fields of the SFE object, and they all have the
same arguments except for arguments specific to the field of the SFE
object. The argument <code>type</code>, which indicates the
corresponding function names in <code>spdep</code>, determines which
spatial statistics are computed.</p>
<p>All univariate global methods in <code>Voyager</code> are listed
here:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"uni"</span>, scope <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="co">#&gt;              name                                           description</span></span>
<span><span class="co">#&gt; 1           moran                                             Moran's I</span></span>
<span><span class="co">#&gt; 2           geary                                             Geary's C</span></span>
<span><span class="co">#&gt; 3        moran.mc                    Moran's I with permutation testing</span></span>
<span><span class="co">#&gt; 4        geary.mc                    Geary's C with permutation testing</span></span>
<span><span class="co">#&gt; 5    sp.mantel.mc Mantel-Hubert spatial general cross product statistic</span></span>
<span><span class="co">#&gt; 6      moran.test                                        Moran's I test</span></span>
<span><span class="co">#&gt; 7      geary.test                                        Geary's C test</span></span>
<span><span class="co">#&gt; 8    globalG.test                                         Global G test</span></span>
<span><span class="co">#&gt; 9  sp.correlogram                                           Correlogram</span></span>
<span><span class="co">#&gt; 10      variogram                                  Variogram with model</span></span>
<span><span class="co">#&gt; 11  variogram_map                                         Variogram map</span></span></code></pre></div>
<p>When calling <code>calculate*variate()</code> or
<code>run*variate()</code>, the <code>type</code> (2nd) argument takes a
string that matches an entry in the <code>name</code> column in the data
frame returned by <code><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods()</a></code>.</p>
<p>To demonstrate spatial autocorrelation in gene expression, top highly
variable genes (HVGs) are used. The HVGs are found with the
<code>scran</code> method.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>A global statistic yields one result for the entire dataset.</p>
<div class="section level4">
<h4 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h4>
<p>As a reference, Pearson correlation is defined as</p>
<p><span class="math display">\[
\rho_{X,Y} = \frac{\sum_{i=1}^n\sum_{j=1}^n (x_i - \bar x)(y_i - \bar
y)}{\sqrt{\sum_{i=1}^n (x_i - \bar x)^2}\sqrt{\sum_{i=1}^n (y_i - \bar
y)^2}}.
\]</span></p>
<p>There are several ways to quantify spatial autocorrelation, the most
common of which is Moran’s I <span class="citation">(Moran
1950)</span>:</p>
<p><span class="math display">\[
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i -
\bar{x})^2},
\]</span></p>
<p>where <span class="math inline">\(n\)</span> is the number of spots
or locations, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are different locations, or spots in
the Visium context, <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> are variables with values at each
location, and <span class="math inline">\(w_{ij}\)</span> is a spatial
weight, which can be inversely proportional to distance between spots or
an indicator of whether two spots are neighbors, subject to various
definitions of neighborhood and whether to normalize the number of
neighbors. The <a href="https://r-spatial.github.io/spdep/index.html" class="external-link"><code>spdep</code></a>
package uses the neighborhood.</p>
<p>Moran’s I is analogous to the Pearson correlation between the value
at each location and the average value at its neighbors (but not
identical, see <span class="citation">(Lee 2001)</span>). Just like
Pearson correlation, Moran’s I is generally bound between -1 and 1,
where positive value indicates positive spatial autocorrelation,
i.e. nearby values tend to be more similar, and negative value indicates
negative spatial autocorrelation, i.e. nearby values tend to be more
dissimilar.</p>
<p>Upon visual inspection earlier in the workshop, total UMI counts per
spot (nCounts) seem to have spatial autocorrelation. For numeric columns
of <code>colData(sfe)</code>, all univariate methods can be called with
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colDataUnivariate()</a></code>. Here we compute Moran’s I for nCounts
and nGenes:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"moran"</span>, </span>
<span>                                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,</span>
<span>                                colGraphName <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;         moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts    0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes     0.384028   3.88036</span></span></code></pre></div>
<p>For <code>colData</code>, the results are added to
<code>colFeatureData(sfe)</code>, and features for which Moran’s I is
not calculated have NA. The column names of <code>featureData</code>
distinguishes between different samples (there’s only one sample in this
dataset), and are parsed by plotting functions. Here the first column is
the Moran’s I value, which indicates moderate positive spatial
autocorrelation for both nCounts and nGenes. The second column is
kurtosis of the data.</p>
<p>Compute Moran’s I for attributes of geometries: Here “area” is the
area of the cross section of each myofiber as seen in this tissue
section and “eccentricity” is the eccentricity of the ellipse fitted to
each myofiber.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remember zero.policy = TRUE since there're singletons</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">annotGeometryUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"moran"</span>,</span>
<span>                                      features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="st">"eccentricity"</span><span class="op">)</span>, </span>
<span>                                      annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span>
<span>                                      annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>                                      zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_simplified"</span><span class="op">)</span>, <span class="st">"featureData"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="st">"eccentricity"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;              moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; area            0.327888   4.95675</span></span>
<span><span class="co">#&gt; eccentricity    0.110938   3.26913</span></span></code></pre></div>
<p>For a non-geometry column in a <code>colGeometry</code>,
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colGeometryUnivariate()</a></code> is like
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">annotGeometryUnivariate()</a></code> here, but none of the
<code>colGeometries</code> in this dataset has extra columns.</p>
<p>For gene expression, the <code>logcounts</code> assay is used by
default (use the <code>exprs_values</code> argument to change the
assay), though this may or may not be best practice. If the metrics are
computed for a large number of features, parallel computing is
supported, with <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link"><code>BiocParallel</code></a>,
with the <code>BPPARAM</code> argument.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"moran"</span>, features <span class="op">=</span> <span class="va">hvgs</span>, </span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"moran_Vis5A"</span>, <span class="st">"K_Vis5A"</span>, <span class="st">"symbol"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">#&gt;                    moran_Vis5A   K_Vis5A      symbol</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000029304    0.734937   1.63516        Spp1</span></span>
<span><span class="co">#&gt; ENSMUSG00000050708    0.665563   1.81841        Ftl1</span></span>
<span><span class="co">#&gt; ENSMUSG00000050335    0.741474   1.68098      Lgals3</span></span>
<span><span class="co">#&gt; ENSMUSG00000021939    0.708362   1.86896        Ctsb</span></span>
<span><span class="co">#&gt; ENSMUSG00000021190    0.659916   1.66838        Lgmn</span></span>
<span><span class="co">#&gt; ENSMUSG00000018893    0.675840   1.82510          Mb</span></span></code></pre></div>
<p>As Moran’s I is very commonly used,
<code>runMoransI(sfe_tissue, features = hvgs)</code> is equivalent to
<code>runUnivariate(sfe_tissue, type = "moran", features = hvgs)</code>.</p>
</div>
<div class="section level4">
<h4 id="exercises-2">Exercises<a class="anchor" aria-label="anchor" href="#exercises-2"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Use <code><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods()</a></code> to find the “name” of Geary’s C
<span class="citation">(Geary 1954)</span>. This name should be used in
the <code>type</code> argument in <code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate()</a></code>.</li>
<li>Compute Geary’s C on the highly variable genes, and show the
results. Interpretation of Geary’s C: a value below 1 indicates positive
spatial autocorrelation, while a value above 1 indicates negative
spatial autocorrelation.</li>
</ol>
</div>
<div class="section level4">
<h4 id="further-reading">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Spatial transcriptomics data is usually much larger than the typical
geospatial dataset back in the 1950s when Moran’s I and Geary’s C were
devised. See <a href="https://link.springer.com/article/10.1007/s10109-019-00293-3" class="external-link"><span class="citation">(Luo, Griffith, and Wu 2019)</span></a> for asymptotic
properties of Moran’s I for large datasets with normal and skewed
distributions.</li>
<li>The negative binomial distribution is often used to model
transcriptomics data due to bursts in transcription, although the
Poisson distribution is sometimes used instead to simplify the math. See
<a href="https://onlinelibrary.wiley.com/doi/10.1111/j.0016-7363.2006.00679.x" class="external-link"><span class="citation">(Griffith and Haining 2006)</span></a> for a
consideration of the Poisson distribution in spatial analyses.</li>
<li>Moran’s I is not exactly the same as Pearson correlation between the
values themselves and the spatially smoothed values. The bounds of
Moran’s I depend on the spatial neighborhood graph. Usually the upper
bound is around 1, while the lower bound is closer to -0.5 than -1. See
<a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1538-4632.1984.tb00797.x" class="external-link"><span class="citation">(Jong, Sprenger, and Veen 1984)</span></a> for a
derivation of extreme values of Moran’s I and Geary’s C.</li>
<li>Spatial autocorrelation decays at different length scales for
different features, and the correlogram is one way to find the length
scales. <a href="https://pachterlab.github.io/voyager/articles/correlogram_landing.html" class="external-link">These
are the vignettes</a> that use correlograms. Also see <a href="https://pachterlab.github.io/voyager/articles/vig6_merfish.html" class="external-link">this
vignette</a> on Moran’s I flipping signs at different length
scales.</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="univariate-local">Univariate local<a class="anchor" aria-label="anchor" href="#univariate-local"></a>
</h3>
<p>Local statistics yield a result at each location rather than the
whole dataset, while global statistics may obscure local heterogeneity.
See <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/j.1538-4632.2009.00767.x" class="external-link"><span class="citation">(Fotheringham 2009)</span></a> for an interesting
discussion of relationships between global and local spatial statistics.
Local statistics are stored in the <code>localResults</code> field of
the SFE object, which can be accessed by the <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult()</a></code>
or <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResults()</a></code> functions in the
<code>SpatialFeatureExperiment</code> package.</p>
<p>All univariate local methods in <code>Voyager</code> are listed
here:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"uni"</span>, scope <span class="op">=</span> <span class="st">"local"</span><span class="op">)</span></span>
<span><span class="co">#&gt;               name                                          description</span></span>
<span><span class="co">#&gt; 1       localmoran                                      Local Moran's I</span></span>
<span><span class="co">#&gt; 2  localmoran_perm                  Local Moran's I permutation testing</span></span>
<span><span class="co">#&gt; 3           localC                                      Local Geary's C</span></span>
<span><span class="co">#&gt; 4      localC_perm                  Local Geary's C permutation testing</span></span>
<span><span class="co">#&gt; 5           localG                                      Getis-Ord Gi(*)</span></span>
<span><span class="co">#&gt; 6      localG_perm             Getis-Ord Gi(*) with permutation testing</span></span>
<span><span class="co">#&gt; 7             LOSH                     Local spatial heteroscedasticity</span></span>
<span><span class="co">#&gt; 8          LOSH.mc Local spatial heteroscedasticity permutation testing</span></span>
<span><span class="co">#&gt; 9          LOSH.cs     Local spatial heteroscedasticity Chi-square test</span></span>
<span><span class="co">#&gt; 10      moran.plot                                   Moran scatter plot</span></span></code></pre></div>
<div class="section level4">
<h4 id="local-morans-i">Local Moran’s I<a class="anchor" aria-label="anchor" href="#local-morans-i"></a>
</h4>
<p>To recap, global Moran’s I is defined as</p>
<p><span class="math display">\[
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i -
\bar{x})^2}.
\]</span></p>
<p>Local Moran’s I <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1538-4632.1995.tb00338.x" class="external-link"><span class="citation">(Anselin 1995)</span></a> is defined as</p>
<p><span class="math display">\[
I_i = (n-1)\frac{(x_i - \bar{x})\sum_{j=1}^n w_{ij} (x_j -
\bar{x})}{\sum_{i=1}^n (x_i - \bar{x})^2}.
\]</span></p>
<p>It’s similar to global Moran’s I, but the values at locations <span class="math inline">\(i\)</span> are not summed and there’s no
normalization by the sum of spatial weights. Local Moran’s I has been
used in spatial transcriptomics in the MERINGUE package <span class="citation">(Miller et al. 2021)</span>. Here we compute local
Moran’s I for the gene Myh2.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>,</span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p>It is useful to plot the log normalized Myh2 gene expression as
context to interpret the local results:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                   swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                   linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<p>Any local spatial results can be plotted with
<code>plotLocalResults()</code>, which is similar to
<code><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature()</a></code>. Here a divergent palette is used
because Moran’s I has a sensible center at 0 (actually the expected
value of Moran’s I is -1/(n-1) when the mean is unknown, but it’s very
close to 0 as n is typically large in spatial -omics).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
<p>We see that myofiber regions with higher Myh2 expression also have
stronger spatial autocorrelation, while the injury site locally has some
negative spatial autocorrelation.</p>
<p>The results are stored in the <code>localResults</code> field in the
SFE object, with getters and setters analogous to
<code>reducedDims</code>, but the name of the local method and the
feature/gene for which the local method was run need to be specified as
well.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"localmoran"</span>, feature <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                  swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lr</span><span class="op">)</span></span>
<span><span class="co">#&gt;                          Ii          E.Ii     Var.Ii       Z.Ii Pr(z != E(Ii))</span></span>
<span><span class="co">#&gt; AAACATTTCCCGGATT 2.12545883 -0.0012237995 0.37891181 3.45488511   5.505274e-04</span></span>
<span><span class="co">#&gt; AAACCTAAGCAGCCGG 3.33088903 -0.0038553468 0.59334776 4.32920244   1.496503e-05</span></span>
<span><span class="co">#&gt; AAACGAGACGGTTGAT 0.30430735 -0.0009817045 0.15152269 0.78428227   4.328745e-01</span></span>
<span><span class="co">#&gt; AAACGGGCGTACGGGT 4.69775712 -0.0063342069 0.97242481 4.77032239   1.839313e-06</span></span>
<span><span class="co">#&gt; AAACTCGGTTCGCAAT 0.01991573 -0.0002817611 0.04351933 0.09681804   9.228709e-01</span></span>
<span><span class="co">#&gt; AAACTGCTGGCTCCAA 0.55285063 -0.0009817045 0.15152269 1.42278566   1.547983e-01</span></span>
<span><span class="co">#&gt;                       mean    median     pysal    -log10p -log10p_adj</span></span>
<span><span class="co">#&gt; AAACATTTCCCGGATT High-High High-High High-High 3.25922109    2.657161</span></span>
<span><span class="co">#&gt; AAACCTAAGCAGCCGG High-High High-High High-High 4.82492235    3.979824</span></span>
<span><span class="co">#&gt; AAACGAGACGGTTGAT   Low-Low   Low-Low   Low-Low 0.36363800    0.000000</span></span>
<span><span class="co">#&gt; AAACGGGCGTACGGGT High-High High-High High-High 5.73534434    4.890246</span></span>
<span><span class="co">#&gt; AAACTCGGTTCGCAAT High-High High-High High-High 0.03485905    0.000000</span></span>
<span><span class="co">#&gt; AAACTGCTGGCTCCAA   Low-Low   Low-Low   Low-Low 0.81023381    0.000000</span></span></code></pre></div>
<p>It is interesting to see how spatial autocorrelation relates to gene
expression level, much as finding how variance relates to mean in the
expression of each gene, which usually indicates overdispersion compared
to Poisson in scRNA-seq and Visium data:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>myh2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span> <span class="op">==</span> <span class="st">"Myh2"</span>,<span class="op">]</span>,</span>
<span>                 Ii <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="st">"Myh2"</span>, </span>
<span>                                  swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Ii"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">myh2</span>, <span class="va">Ii</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">myh2</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Myh2 (log counts)"</span>, y <span class="op">=</span> <span class="st">"localmoran"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<p>For this gene, Visium spots with higher expression also tend to have
higher local Moran’s I, but this may or may not apply to other genes.
The vertical dashed line marks the mean gene expression; note the
subtraction of mean in the expression for global and local Moran’s I,
leading spots close to mean to have local Moran’s I close to 0.</p>
<p>Local spatial analyses often return a matrix or data frame. The
<code><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult()</a></code> function has a default column for each
local spatial method, but other columns can be plotted as well. Use the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs()</a></code> function to see which columns are
present, and use the <code>attribute</code> argument to specify which
column to plot.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="st">"Myh2"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Ii"             "E.Ii"           "Var.Ii"         "Z.Ii"          </span></span>
<span><span class="co">#&gt;  [5] "Pr(z != E(Ii))" "mean"           "median"         "pysal"         </span></span>
<span><span class="co">#&gt;  [9] "-log10p"        "-log10p_adj"</span></span></code></pre></div>
<p>Some local spatial methods return p-values at each location, in a
column with name like <code>Pr(z != E(Ii))</code>, where the test is two
sided (default, can be changed with the <code>alternative</code>
argument in <code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate()</a></code> which is passed to the relevant
underlying function in <code>spdep</code>). Negative log of the p-value
is computed to facilitate visualization (smaller or more significant
p-values are plotted as higher values), and the p-value is corrected for
multiple hypothesis testing with <code>p.adjustSP()</code> in
<code>spdep</code>, where the number of tests is the number of neighbors
of each location rather than the total number of locations
(<code>-log10p_adj</code>).</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, attribute <span class="op">=</span> <span class="st">"-log10p_adj"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<p>In this plot and all following plots of p-values, a divergent palette
is used to show locations that are significant after adjusting for
multiple testing and those that are not significant in different colors.
The center of the divergent palette is p = 0.05, so the brown spots are
significant while a dark blue means <em>really</em> not significant.</p>
<p>The “pysal” column shows the type of neighborhood, such as whether
low value is near other low values, or high value is near other high
values.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, attribute <span class="op">=</span> <span class="st">"pysal"</span>, </span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="exercises-3">Exercises<a class="anchor" aria-label="anchor" href="#exercises-3"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Compute local spatial heteroscedasticity (LOSH) <span class="citation">(J. Keith Ord and Getis 2012)</span> on Myh2 and plot
the results. A sequential palette is appropriate.</li>
<li>Which other columns are returned by LOSH? Plot one of them in space.
See documentation of <code><a href="https://r-spatial.github.io/spdep/reference/LOSH.html" class="external-link">spdep::LOSH()</a></code> for the meanings of the
other columns.</li>
<li>How does the spatial pattern of LOSH compare to that of local Moran
for the same genes?</li>
</ol>
</div>
<div class="section level4">
<h4 id="further-reading-1">Further reading<a class="anchor" aria-label="anchor" href="#further-reading-1"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Getis-Ord Gi* <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1538-4632.1995.tb00912.x" class="external-link"><span class="citation">(J. K. Ord and Getis 1995)</span></a> is another
commonly used local spatial statistic which identifies hotspots (high
values clustered together in space) and coldspots (low values clostered
in space). <a href="https://pachterlab.github.io/voyager/articles/getisord_landing.html" class="external-link">These
vignettes</a> use Getis-Ord Gi*.</li>
<li>The Moran scatter plot <a href="https://www.taylorfrancis.com/chapters/edit/10.1201/9780203739051-8/moran-scatterplot-esda-tool-assess-local-instability-spatial-association-luc-anselin" class="external-link"><span class="citation">(Anselin 1996)</span></a> is another ESDA tool.
<code>Voyager</code> has a special function to plot the results
<code><a href="https://rdrr.io/pkg/Voyager/man/moranPlot.html" class="external-link">moranPlot()</a></code>. See <a href="https://pachterlab.github.io/voyager/articles/moranplot_landing.html" class="external-link">these
vignettes</a> for examples of Moran scatter plot applied to spatial
transcriptomics.</li>
<li>Analogous to the Moran scatter plot is the Geary scatter plot (not
yet implemented in <code>Voyager</code>) proposed in <a href="https://link.springer.com/article/10.1007/s43071-022-00031-w" class="external-link"><span class="citation">(Griffith and Chun 2022)</span></a> which is said to
better detect local negative spatial autocorrelation. This paper also
includes other considerations on Moran’s I and Geary’s C.</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="bivariate">Bivariate<a class="anchor" aria-label="anchor" href="#bivariate"></a>
</h3>
<p>Some spatial methods analyze how two variables relate. A list of all
bivariate global methods can be seen here:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"bi"</span>, scope <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  name                                     description</span></span>
<span><span class="co">#&gt; 1                 lee                       Lee's bivariate statistic</span></span>
<span><span class="co">#&gt; 2              lee.mc Lee's bivariate static with permutation testing</span></span>
<span><span class="co">#&gt; 3            lee.test                                    Lee's L test</span></span>
<span><span class="co">#&gt; 4     cross_variogram                                 Cross variogram</span></span>
<span><span class="co">#&gt; 5 cross_variogram_map                             Cross variogram map</span></span></code></pre></div>
<p>There are also local bivariate methods:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"bi"</span>, scope <span class="op">=</span> <span class="st">"local"</span><span class="op">)</span></span>
<span><span class="co">#&gt;            name                     description</span></span>
<span><span class="co">#&gt; 1      locallee Local Lee's bivariate statistic</span></span>
<span><span class="co">#&gt; 2 localmoran_bv       Local bivariate Moran's I</span></span></code></pre></div>
<div class="section level4">
<h4 id="lees-l">Lee’s L<a class="anchor" aria-label="anchor" href="#lees-l"></a>
</h4>
<p>Lee’s L <a href="https://link.springer.com/article/10.1007/s101090100064" class="external-link"><span class="citation">(Lee 2001)</span></a> was developed from relating
Moran’s I to Pearson correlation, and is defined as</p>
<p><span class="math display">\[
L_{X,Y} = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\left[ \sum_{j=1}^n w_{ij}  (x_j - \bar{x}) \right] \left[ \sum_{j=1}^n
w_{ij} (y_j - \bar{y}) \right]}{\sqrt{\sum_{i=1}^n (x_i -
\bar{x})^2}\sqrt{\sum_{i=1}^n (y_i - \bar{y})^2} },
\]</span></p>
<p>where <span class="math inline">\(n\)</span> is the number of spots
or locations, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are different locations, or spots in
the Visium context, <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> are variables with values at each
location, and <span class="math inline">\(w_{ij}\)</span> is a spatial
weight, which can be inversely proportional to distance between spots or
an indicator of whether two spots are neighbors, subject to various
definitions of neighborhood. The <code>Giotto</code> package has
implemented something like Lee’s L <span class="citation">(Dries et al.
2021)</span>.</p>
<p>Here we compute Lee’s L for top highly variagle genes (HVGs) in this
dataset:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, fdr.threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p>Because bivariate global results can have very different formats
(matrix for Lee’s L and lists for many other methods), the results are
not stored in the SFE object. The <code><a href="https://rdrr.io/pkg/Voyager/man/calculateBivariate.html" class="external-link">calculateBivariate()</a></code>
function is used to perform all bivariate analyses. Analogous to
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate()</a></code> there is <code><a href="https://rdrr.io/pkg/Voyager/man/calculateBivariate.html" class="external-link">runBivariate()</a></code> which
stores the results in the SFE object, but it only applies to local
bivariate methods whose results have a more uniform format and are
stored in the <code>localResults</code> field just like local univariate
results.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateBivariate.html" class="external-link">calculateBivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"lee"</span>, feature1 <span class="op">=</span> <span class="va">hvgs</span><span class="op">)</span></span></code></pre></div>
<p>This gives a spatially informed correlation matrix among the genes,
which can be plotted as a heatmap:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pal_rng</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/getDivergeRange.html" class="external-link">getDivergeRange</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scico/man/scico.html" class="external-link">scico</a></span><span class="op">(</span><span class="fl">256</span>, begin <span class="op">=</span> <span class="va">pal_rng</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, end <span class="op">=</span> <span class="va">pal_rng</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, palette <span class="op">=</span> <span class="st">"vik"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">res</span>, color <span class="op">=</span> <span class="va">pal</span>, show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>         show_colnames <span class="op">=</span> <span class="cn">FALSE</span>, cellwidth <span class="op">=</span> <span class="fl">1</span>, cellheight <span class="op">=</span> <span class="fl">1</span>,</span>
<span>         treeheight_col <span class="op">=</span> <span class="fl">0</span>, treeheight_row <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-44-1.png" width="528"></p>
<p>Some coexpression blocks can be seen. Note that unlike in Pearson
correlation, the diagonal is not 1, because</p>
<p><span class="math display">\[
L_{X,X} = \frac{\sum_i (\tilde x_i - \bar x)^2}{\sum_i (x_i - \bar x)^2}
= \mathrm{SSS}_X,
\]</span></p>
<p>which is approximated the ratio between the variance of spatially
lagged <span class="math inline">\(x\)</span> and variance of <span class="math inline">\(x\)</span>. Because the spatial lag introduces
smoothing, the spatial lag reduced variance, making the diagonal less
than 1. This is the spatial smoothing scalar (SSS), and Moran’s I is
approximately Pearson correlation between <span class="math inline">\(X\)</span> and spatially lagged <span class="math inline">\(X\)</span> (<span class="math inline">\(\tilde
X\)</span>) multiplied by SSS:</p>
<p><span class="math display">\[
I \approx \mathrm{SSS}_X \cdot \rho_{X, \tilde X}
\]</span></p>
<p>Similarly for Lee’s L, as shown in <span class="citation">(Lee
2001)</span>,</p>
<p><span class="math display">\[
L_{X, Y} = \sqrt{\mathrm{SSS}_X}\sqrt{\mathrm{SSS}_Y} \cdot \rho_{\tilde
X, \tilde Y}
\]</span></p>
<p>With more spatial clustering, the variance is less reduced by the
spatial lag, leading to a larger SSS. Hence when both <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> are spatially distributed like salt and
pepper while strongly correlated, Lee’s L will be low because the lack
of spatial autocorrelation leads to a small SSS.</p>
<p>Weighted correlation network analysis (WGCNA) <span class="citation">(Langfelder and Horvath 2008)</span> is a time honored
method to find gene co-expression modules, and it can take any
correlation matrix. Then it would be interesting to apply WGCNA to the
Lee’s L matrix to identify spatially informed gene co-expression
modules.</p>
</div>
<div class="section level4">
<h4 id="exercises-4">Exercises<a class="anchor" aria-label="anchor" href="#exercises-4"></a>
</h4>
<p>Local Lee’s L is analogous to local Moran’s I – a disaggregated form
of Lee’s L showing the contribution of each spot to the global Lee’s L.
This is derived in <span class="citation">(Lee 2001)</span> after global
Lee’s L is shown.</p>
<p>Run local Lee’s L on two genes of your choice. You can use the
myofiber type marker genes Myh7, Myh2, Myh4, and Myh1. Then plot the
results in space. Hint: Use <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultFeatures()</a></code> to find
what name the results are stored under. How would you interpret the
results?</p>
<p>See <a href="https://pachterlab.github.io/voyager/articles/bivariate.html" class="external-link">this
vignette</a> on bivariate methods in Voyager applied to this dataset,
including local Lee’s L and a bivariate version of local Moran’s I.</p>
</div>
</div>
<div class="section level3">
<h3 id="multivariate">Multivariate<a class="anchor" aria-label="anchor" href="#multivariate"></a>
</h3>
<p>Spatial transcriptomics data can have anywhere from hundreds of genes
to the whole genome. It would be tedious to manually check univariate
spatial statistics one gene at a time. Furthermore, genes are often
co-regulated, while univariate spatial statistics is blind to the
co-regulation. Hence we have multivariate spatial statistics, analyzing
multiple genes simultaneously while taking spatial information into
account. Multivariate spatial methods in <code>Voyager</code> are listed
here:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/listSFEMethods.html" class="external-link">listSFEMethods</a></span><span class="op">(</span><span class="st">"multi"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                name                                      description</span></span>
<span><span class="co">#&gt; 1        multispati                                   MULTISPATI PCA</span></span>
<span><span class="co">#&gt; 2      localC_multi                     Multivariate local Geary's C</span></span>
<span><span class="co">#&gt; 3 localC_perm_multi Multivariate local Geary's C permutation testing</span></span></code></pre></div>
<div class="section level4">
<h4 id="non-spatial-pca">Non-spatial PCA<a class="anchor" aria-label="anchor" href="#non-spatial-pca"></a>
</h4>
<p>First we run regular principal component analysis (PCA), to compare
to a type of spatially informed PCA known as MULTISPATI PCA <span class="citation">(Stéphane Dray, Saı̈d, and Débias 2008)</span>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvgs2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, subset_row <span class="op">=</span> <span class="va">hvgs2</span>,</span>
<span>                     exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Use the elbow plot to see variance explained by each PC:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/ElbowPlot.html" class="external-link">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-48-1.png" width="700"></p>
<p>Plot top gene loadings in each PC, which are the contribution of each
gene to each PC:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotDimLoadings.html" class="external-link">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-49-1.png" width="700"></p>
<p>Plot the first 4 PCs in space</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/spatialReducedDim.html" class="external-link">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span>, <span class="fl">4</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-50-1.png" width="768"></p>
<p>The first PC separates the leukocyte infiltrated injury site from the
myofibers, while PC2 and PC3 tease out the muscle tendon junctions.</p>
</div>
<div class="section level4">
<h4 id="multispati-pca">MULTISPATI PCA<a class="anchor" aria-label="anchor" href="#multispati-pca"></a>
</h4>
<p>Spatially informed dimension reduction is actually not new, and dates
back to at least 1985, with Wartenberg’s crossover of Moran’s I and PCA
<a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1538-4632.1985.tb00849.x" class="external-link"><span class="citation">(Wartenberg 1985)</span></a>, which was generalized and
further developed as MULTISPATI PCA <a href="https://onlinelibrary.wiley.com/doi/10.3170/2007-8-18312" class="external-link"><span class="citation">(Stéphane Dray, Saı̈d, and Débias 2008)</span></a>,
implemented in the <a href="https://cran.r-project.org/web/packages/adespatial/index.html" class="external-link"><code>adespatial</code></a>
package on CRAN. In short, while PCA tries to maximize the variance
explained by each PC, MULTISPATI maximizes the product of Moran’s I and
variance explained. Also, while all the eigenvalues from PCA are
non-negative, because the covariance matrix is positive semidefinite,
MULTISPATI can give negative eigenvalues, which represent negative
spatial autocorrelation, which can be present and interesting but is not
as common as positive spatial autocorrelation and is often masked by the
latter <a href="https://www.mdpi.com/2571-905X/2/3/27" class="external-link"><span class="citation">(Griffith 2019)</span></a>.</p>
<p>In single cell -omics conventions, let <span class="math inline">\(X\)</span> denote a gene count matrix whose
columns are cells or Visium spots and whose rows are genes, with <span class="math inline">\(n\)</span> columns. Let <span class="math inline">\(W\)</span> denote the row normalized <span class="math inline">\(n\times n\)</span> adjacency matrix of the spatial
neighborhood graph of the cells or Visium spots, which does not have to
be symmetric. MULTISPATI diagonalizes a symmetric matrix</p>
<p><span class="math display">\[
H = \frac 1 {2n} X(W^t+W)X^t
\]</span></p>
<p>However, the implementation in <code>adespatial</code> is more
general and can be used for other multivariate analyses in the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3265363/" class="external-link">duality
diagram</a> paradigm, such as correspondence analysis; the equation
above is simplified just for PCA, without having to introduce the
duality diagram here.</p>
<p>Here we compute MULTISPATI PCA, with the 20 most positive and 20 most
negative eigenvalues.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMultivariate.html" class="external-link">runMultivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"multispati"</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>,</span>
<span>                              nfposi <span class="op">=</span> <span class="fl">20</span>, nfnega <span class="op">=</span> <span class="fl">20</span>, subset_row <span class="op">=</span> <span class="va">hvgs2</span><span class="op">)</span></span></code></pre></div>
<p>Then plot the most positive and most negative eigenvalues. Note that
the eigenvalues here are not variance explained. Instead, they are the
product of variance explained and Moran’s I. So the most positive
eigenvalues correspond to eigenvectors that simultaneously explain more
variance and have large positive Moran’s I. The most negative
eigenvalues correspond to eigenvectors that simultaneously explain more
variance and have negative Moran’s I.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/ElbowPlot.html" class="external-link">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, nfnega <span class="op">=</span> <span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"multispati"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-52-1.png" width="700"></p>
<p>Here the positive eigenvalues drop sharply after, PC1, while none of
the negative eigenvalues seem noteworthy. However, in spatial
transcriptomics datasets with single cell resolution, there can be very
negative eigenvalues and the corresponding PC is biologically relevant,
see <a href="https://pachterlab.github.io/voyager/articles/multispati.html#multispati-pca" class="external-link">this
vignette on mouse liver data</a>.</p>
<p>What do these components mean? Each component is a linear combination
of genes to maximize the product of variance explained and Moran’s I.
The second component maximizes this product provided that it’s
orthogonal to the first component, and so on. As the loss in variance
explained is usually not huge, these components can be considered axes
along which <em>spatially coherent</em> groups of spots are separated
from each other as much as possible according to expression of the
highly variable genes, so in theory, clustering with positive MULTISPATI
components should give more spatially coherent clusters. Because of the
spatial coherence, MULTISPATI might be more robust to outliers.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotDimLoadings.html" class="external-link">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, reduction <span class="op">=</span> <span class="st">"multispati"</span>,</span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-53-1.png" width="700"></p>
<p>Plot the these PCs:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/spatialReducedDim.html" class="external-link">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"multispati"</span>, <span class="fl">4</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                  diverge_center <span class="op">=</span> <span class="fl">0</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-54-1.png" width="768"></p>
<p>Here unlike in non-spatial PCA, PC4 continues to be spatially
structured.</p>
<div class="section level5">
<h5 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h5>
<p>PCA embeddings are often used for clustering in scRNA-seq data
analysis. Here we perform Leiden clustering on non-spatial and
MULTISPATI PCA embeddings.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe_tissue</span><span class="op">$</span><span class="va">clusts_nonspatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                                             BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span></span>
<span>                                                 cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                                 cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                                     objective_function <span class="op">=</span> <span class="st">"modularity"</span>,</span>
<span>                                                     resolution_parameter <span class="op">=</span> <span class="fl">1</span></span>
<span>                                                 <span class="op">)</span></span>
<span>                                             <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>See if clustering with the positive MULTISPATI PCs give more
spatially coherent clusters</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe_tissue</span><span class="op">$</span><span class="va">clusts_multispati</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"multispati"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>, </span>
<span>                                            BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span></span>
<span>                                                cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                                cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                                    objective_function <span class="op">=</span> <span class="st">"modularity"</span>,</span>
<span>                                                    resolution_parameter <span class="op">=</span> <span class="fl">1</span></span>
<span>                                                <span class="op">)</span></span>
<span>                                            <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Plot the clusters in space:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clusts_nonspatial"</span>, <span class="st">"clusts_multispati"</span><span class="op">)</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">2</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-57-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="spatial-autocorrelation-of-principal-components">Spatial autocorrelation of principal components<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-principal-components"></a>
</h5>
<p>Here we compare Moran’s I for cell embeddings in each non-spatial and
MULTISPATI PC. Just like there’s <code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colDataUnivariate()</a></code> or
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colDataMoransI()</a></code> for <code>colData</code> columns and
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">annotGeometryUnivariate()</a></code> for attributes of annotation
geometries, univariate spatial statistics can be computed for cell/spot
embeddings in reduced dimensions, with
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">reducedDimUnivariate()</a></code>, or for Moran’s I,
<code><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">reducedDimMoransI()</a></code>. The arguments of all these functions
are similar.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># non-spatial</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">reducedDimMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, components <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="co"># spatial</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">reducedDimMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dimred <span class="op">=</span> <span class="st">"multispati"</span>, components <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>PCA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_Vis5A</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>,</span>
<span>                   MULTISPATI_pos <span class="op">=</span> </span>
<span>                       <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"multispati"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_Vis5A</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>,</span>
<span>                   MULTISPATI_neg <span class="op">=</span> </span>
<span>                       <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>,<span class="st">"multispati"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_Vis5A</span><span class="op">[</span><span class="fl">21</span><span class="op">:</span><span class="fl">40</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>                       <span class="fu"><a href="https://rspatial.github.io/terra/reference/flip.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                   index <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ditto_colors"</span><span class="op">)</span></span></code></pre></div>
<p>These are the lower and upper bounds of Moran’s I given this spatial
neighborhood graph according to <span class="citation">(Jong, Sprenger,
and Veen 1984)</span>.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Takes a while if not using optimized BLAS</span></span>
<span><span class="op">(</span><span class="va">mb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/moranBounds.html" class="external-link">moranBounds</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Imin       Imax </span></span>
<span><span class="co">#&gt; -0.5762132  1.0021884</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_moran</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">index</span>, values_to <span class="op">=</span> <span class="st">"value"</span>, names_to <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">index</span>, <span class="va">value</span>, color <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">ditto_colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">mb</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_pretty.html" class="external-link">breaks_pretty</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_width.html" class="external-link">breaks_width</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Moran's I"</span>, color <span class="op">=</span> <span class="st">"Type"</span>, x <span class="op">=</span> <span class="st">"Component"</span><span class="op">)</span></span></code></pre></div>
<p><img src="voyager_files/figure-html/unnamed-chunk-62-1.png" width="700"></p>
<p>The lower and upper bounds of Moran’s I are plotted as horizontal
dashed line. In non-spatial PCA, Moran’s I drops from PC1 to PC6, while
Moran’s I remains high for the subsequent MULTISPATI PCs. Given the
lower bound of Moran’s I, the negative PCs have strong negative spatial
autocorrelation. However, this should not be over-interpreted for this
dataset because of the minuscule magnitude of the negative eigenvalues,
which means that these PCs explain very little variance.</p>
</div>
</div>
<div class="section level4">
<h4 id="exercises-5">Exercises<a class="anchor" aria-label="anchor" href="#exercises-5"></a>
</h4>
<p>Thought experiment: suppose you perform standard PCA and MULTISPATI
PCA on your city, such as on buildings, demographics, or when the city
is divided into pixels, on whichever spatial features you find relevant
to your life. What would the principal components look like? How would
MULTISPATI PC’s differ from the standard PC’s?</p>
</div>
<div class="section level4">
<h4 id="further-reading-2">Further reading<a class="anchor" aria-label="anchor" href="#further-reading-2"></a>
</h4>
<ol style="list-style-type: decimal">
<li>The other multivariate spatial method in Voyager as of Bioc 3.17 is
a multivariate generalization of local Geary’s C <a href="https://onlinelibrary.wiley.com/doi/10.1111/gean.12164" class="external-link"><span class="citation">(Anselin 2019)</span></a>. See <a href="https://pachterlab.github.io/voyager/articles/localc.html" class="external-link">this
vignette</a> on its application to this dataset.</li>
<li>Unlike similar EDA packages for spatial -omics data, Voyager is
extensible, so you can make the uniform user interface run other spatial
methods, just like in <a href="https://www.tidymodels.org/learn/develop/models/" class="external-link">Tidymodels</a>.
See <a href="https://pachterlab.github.io/voyager/articles/sfemethod" class="external-link">this
vignette</a> on extending Voyager.</li>
<li>There are spatially informed dimension reduction methods designed
for spatial -omics data, although they tend to only consider positive
spatial autocorrelation. For example, see <a href="https://www.nature.com/articles/s41467-022-34879-1" class="external-link"><span class="citation">(Shang and Zhou 2022)</span></a> and <a href="https://www.nature.com/articles/s41592-021-01343-9" class="external-link"><span class="citation">(Velten et al. 2022)</span></a>.</li>
<li>This paper discusses many other types of multivariate spatial
analyses in ecology, besides MULTISPATI PCA <a href="https://esajournals.onlinelibrary.wiley.com/doi/abs/10.1890/11-1183.1" class="external-link"><span class="citation">(S. Dray et al. 2012)</span></a>.</li>
<li>MULTISPATI PCA can be thought of as in between two extremes. One
extreme is standard PCA, which diagonalizes the covariance matrix. The
other extreme is the Moran eigen map (MEM) <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1541-0064.1996.tb00462.x" class="external-link"><span class="citation">(Griffith 1996)</span></a>, which only uses the spatial
weights matrix, without the data matrix. MEM’s are made by diagonalizing
a double centered spatial weights matrix. The first eigenvector has
values that make the largest possible Moran’s I given the spatial
neighborhood graph. The second eigenvector also maximizes Moran’s I
given that it’s orthogonal to the first eigenvector, and so on. The
eigenvalues are Moran’s I multiplied by a constant. The last eigenvector
has the most negative Moran’s I given the spatial neighborhood graph.
These eigenvectors, or MEM’s, represent spatial structures of different
length scales, which can be selected and used as covariates in
regression to account for spatial autocorrelation, in a procedure called
spatial filtering <span class="citation">(Griffith 2000; Griffith and
Peres-Neto 2006)</span>. See <a href="https://cran.r-project.org/web/packages/adespatial/vignettes/tutorial.html" class="external-link">this
vignette of <code>adespatial</code></a> for examples in ecology.</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="caveats">Caveats<a class="anchor" aria-label="anchor" href="#caveats"></a>
</h3>
<ol style="list-style-type: decimal">
<li>The H&amp;E image plotted behind the geometries can alter perception
of the colors of the geometries.</li>
<li>In the devel version of SFE, 3D data is somewhat supported, as in 3D
transcript spot geometries. However, <code>sf</code> is rather limited
in 3D geometric operations. The operations can only be performed on the
x and y coordinates but not the z coordinates.</li>
<li>This workshop demonstrates ESDA on a single sample. However, studies
often produce multiple biological replica in case and control groups.
The ESDA results can be compared across samples, and hopefully in the
Bioc 3.19 release, computed jointly across samples within the same
treatment group.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2024-02-28 r85999)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] bluster_1.13.0                 BiocSingular_1.19.0           </span></span>
<span><span class="co">#&gt;  [3] BiocNeighbors_1.21.2           pheatmap_1.0.12               </span></span>
<span><span class="co">#&gt;  [5] scico_1.5.0                    tidyr_1.3.1                   </span></span>
<span><span class="co">#&gt;  [7] tibble_3.2.1                   BiocParallel_1.37.0           </span></span>
<span><span class="co">#&gt;  [9] patchwork_1.2.0                scales_1.3.0                  </span></span>
<span><span class="co">#&gt; [11] sf_1.0-15                      Matrix_1.6-5                  </span></span>
<span><span class="co">#&gt; [13] rjson_0.2.21                   scater_1.31.2                 </span></span>
<span><span class="co">#&gt; [15] ggplot2_3.5.0                  scran_1.31.0                  </span></span>
<span><span class="co">#&gt; [17] scuttle_1.13.0                 SFEData_1.5.0                 </span></span>
<span><span class="co">#&gt; [19] Voyager_1.5.0                  SingleCellExperiment_1.25.0   </span></span>
<span><span class="co">#&gt; [21] SummarizedExperiment_1.33.3    Biobase_2.63.0                </span></span>
<span><span class="co">#&gt; [23] GenomicRanges_1.55.3           GenomeInfoDb_1.39.6           </span></span>
<span><span class="co">#&gt; [25] IRanges_2.37.1                 S4Vectors_0.41.3              </span></span>
<span><span class="co">#&gt; [27] BiocGenerics_0.49.1            MatrixGenerics_1.15.0         </span></span>
<span><span class="co">#&gt; [29] matrixStats_1.2.0              SpatialFeatureExperiment_1.5.2</span></span>
<span><span class="co">#&gt; [31] BiocStyle_2.31.0              </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7              filelock_1.0.3           </span></span>
<span><span class="co">#&gt;   [3] R.oo_1.26.0               lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt;   [5] edgeR_4.1.17              lattice_0.22-5           </span></span>
<span><span class="co">#&gt;   [7] magrittr_2.0.3            limma_3.59.3             </span></span>
<span><span class="co">#&gt;   [9] sass_0.4.8                rmarkdown_2.25           </span></span>
<span><span class="co">#&gt;  [11] jquerylib_0.1.4           yaml_2.3.8               </span></span>
<span><span class="co">#&gt;  [13] metapod_1.11.1            sp_2.1-3                 </span></span>
<span><span class="co">#&gt;  [15] RColorBrewer_1.1-3        DBI_1.2.2                </span></span>
<span><span class="co">#&gt;  [17] abind_1.4-5               zlibbioc_1.49.0          </span></span>
<span><span class="co">#&gt;  [19] purrr_1.0.2               R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [21] RCurl_1.98-1.14           rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt;  [23] GenomeInfoDbData_1.2.11   ggrepel_0.9.5            </span></span>
<span><span class="co">#&gt;  [25] irlba_2.3.5.1             terra_1.7-71             </span></span>
<span><span class="co">#&gt;  [27] units_0.8-5               RSpectra_0.16-1          </span></span>
<span><span class="co">#&gt;  [29] dqrng_0.3.2               pkgdown_2.0.7            </span></span>
<span><span class="co">#&gt;  [31] DelayedMatrixStats_1.25.1 codetools_0.2-19         </span></span>
<span><span class="co">#&gt;  [33] DropletUtils_1.23.0       DelayedArray_0.29.8      </span></span>
<span><span class="co">#&gt;  [35] tidyselect_1.2.0          farver_2.1.1             </span></span>
<span><span class="co">#&gt;  [37] ScaledMatrix_1.11.0       viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [39] BiocFileCache_2.11.1      jsonlite_1.8.8           </span></span>
<span><span class="co">#&gt;  [41] e1071_1.7-14              systemfonts_1.0.5        </span></span>
<span><span class="co">#&gt;  [43] tools_4.4.0               ggnewscale_0.4.10        </span></span>
<span><span class="co">#&gt;  [45] ragg_1.2.7                Rcpp_1.0.12              </span></span>
<span><span class="co">#&gt;  [47] glue_1.7.0                gridExtra_2.3            </span></span>
<span><span class="co">#&gt;  [49] SparseArray_1.3.4         xfun_0.42                </span></span>
<span><span class="co">#&gt;  [51] EBImage_4.45.0            dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [53] HDF5Array_1.31.6          withr_3.0.0              </span></span>
<span><span class="co">#&gt;  [55] BiocManager_1.30.22       fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [57] boot_1.3-30               rhdf5filters_1.15.2      </span></span>
<span><span class="co">#&gt;  [59] fansi_1.0.6               spData_2.3.0             </span></span>
<span><span class="co">#&gt;  [61] digest_0.6.34             rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [63] mime_0.12                 R6_2.5.1                 </span></span>
<span><span class="co">#&gt;  [65] textshaping_0.3.7         colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [67] wk_0.9.1                  jpeg_0.1-10              </span></span>
<span><span class="co">#&gt;  [69] RSQLite_2.3.5             R.methodsS3_1.8.2        </span></span>
<span><span class="co">#&gt;  [71] utf8_1.2.4                generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [73] data.table_1.15.2         class_7.3-22             </span></span>
<span><span class="co">#&gt;  [75] httr_1.4.7                htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt;  [77] S4Arrays_1.3.4            spdep_1.3-3              </span></span>
<span><span class="co">#&gt;  [79] pkgconfig_2.0.3           gtable_0.3.4             </span></span>
<span><span class="co">#&gt;  [81] blob_1.2.4                XVector_0.43.1           </span></span>
<span><span class="co">#&gt;  [83] htmltools_0.5.7           bookdown_0.38            </span></span>
<span><span class="co">#&gt;  [85] fftwtools_0.9-11          png_0.1-8                </span></span>
<span><span class="co">#&gt;  [87] SpatialExperiment_1.13.0  knitr_1.45               </span></span>
<span><span class="co">#&gt;  [89] curl_5.2.1                proxy_0.4-27             </span></span>
<span><span class="co">#&gt;  [91] cachem_1.0.8              rhdf5_2.47.4             </span></span>
<span><span class="co">#&gt;  [93] stringr_1.5.1             BiocVersion_3.19.1       </span></span>
<span><span class="co">#&gt;  [95] KernSmooth_2.23-22        parallel_4.4.0           </span></span>
<span><span class="co">#&gt;  [97] vipor_0.4.7               AnnotationDbi_1.65.2     </span></span>
<span><span class="co">#&gt;  [99] desc_1.4.3                s2_1.1.6                 </span></span>
<span><span class="co">#&gt; [101] pillar_1.9.0              grid_4.4.0               </span></span>
<span><span class="co">#&gt; [103] vctrs_0.6.5               dbplyr_2.4.0             </span></span>
<span><span class="co">#&gt; [105] beachmat_2.19.1           sfheaders_0.4.4          </span></span>
<span><span class="co">#&gt; [107] cluster_2.1.6             beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt; [109] evaluate_0.23             magick_2.8.3             </span></span>
<span><span class="co">#&gt; [111] cli_3.6.2                 locfit_1.5-9.9           </span></span>
<span><span class="co">#&gt; [113] compiler_4.4.0            rlang_1.1.3              </span></span>
<span><span class="co">#&gt; [115] crayon_1.5.2              labeling_0.4.3           </span></span>
<span><span class="co">#&gt; [117] classInt_0.4-10           fs_1.6.3                 </span></span>
<span><span class="co">#&gt; [119] ggbeeswarm_0.7.2          stringi_1.8.3            </span></span>
<span><span class="co">#&gt; [121] viridisLite_0.4.2         deldir_2.0-4             </span></span>
<span><span class="co">#&gt; [123] munsell_0.5.0             Biostrings_2.71.2        </span></span>
<span><span class="co">#&gt; [125] tiff_0.1-12               ExperimentHub_2.11.1     </span></span>
<span><span class="co">#&gt; [127] sparseMatrixStats_1.15.0  bit64_4.0.5              </span></span>
<span><span class="co">#&gt; [129] Rhdf5lib_1.25.1           KEGGREST_1.43.0          </span></span>
<span><span class="co">#&gt; [131] statmod_1.5.0             highr_0.10               </span></span>
<span><span class="co">#&gt; [133] AnnotationHub_3.11.1      igraph_2.0.2             </span></span>
<span><span class="co">#&gt; [135] memoise_2.0.1             bslib_0.6.1              </span></span>
<span><span class="co">#&gt; [137] bit_4.0.5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Anselin1995-cs" class="csl-entry">
Anselin, Luc. 1995. <span>“Local Indicators of Spatial
<span>Association—LISA</span>.”</span> <em>Geogr. Anal.</em> 27 (2):
93–115.
</div>
<div id="ref-Anselin1996-mo" class="csl-entry">
———. 1996. <span>“The Moran Scatterplot as an <span>ESDA</span> Tool to
Assess Local Instability in Spatial Association.”</span> In <em>Spatial
Analytical Perspectives on <span>GIS</span></em>, 111–26. Routledge.
</div>
<div id="ref-Anselin2019-uv" class="csl-entry">
———. 2019. <span>“A Local Indicator of Multivariate Spatial Association:
Extending Geary’s c.”</span> <em>Geogr. Anal.</em> 51 (2): 133–50.
</div>
<div id="ref-Dray2012-on" class="csl-entry">
Dray, S, R Pélissier, P Couteron, M-J Fortin, P Legendre, P R
Peres-Neto, E Bellier, et al. 2012. <span>“Community Ecology in the Age
of Multivariate Multiscale Spatial Analysis.”</span> <em>Ecol.
Monogr.</em> 82 (3): 257–75.
</div>
<div id="ref-Dray2008-en" class="csl-entry">
Dray, Stéphane, Sonia Saı̈d, and Françis Débias. 2008. <span>“Spatial
Ordination of Vegetation Data Using a Generalization of Wartenberg’s
Multivariate Spatial Correlation.”</span> <em>J. Veg. Sci.</em> 19 (1):
45–56.
</div>
<div id="ref-Dries2021-eh" class="csl-entry">
Dries, Ruben, Qian Zhu, Rui Dong, Chee-Huat Linus Eng, Huipeng Li, Kan
Liu, Yuntian Fu, et al. 2021. <span>“Giotto: A Toolbox for Integrative
Analysis and Visualization of Spatial Expression Data.”</span>
<em>Genome Biol.</em> 22 (1): 78.
</div>
<div id="ref-Fotheringham2009-ak" class="csl-entry">
Fotheringham, A Stewart. 2009. <span>“<span>‘The Problem of Spatial
Autocorrelation’</span> and Local Spatial Statistics.”</span> <em>Geogr.
Anal.</em> 41 (4): 398–403.
</div>
<div id="ref-Geary1954-fq" class="csl-entry">
Geary, R C. 1954. <span>“The Contiguity Ratio and Statistical
Mapping.”</span> <em>The Incorporated Statistician</em> 5 (3): 115–46.
</div>
<div id="ref-Griffith1996-gg" class="csl-entry">
Griffith, Daniel A. 1996. <span>“<span>SPATIAL</span>
<span>AUTOCORRELATION</span> and <span>EIGENFUNCTIONS</span>
<span>OF</span> <span>THE</span> <span>GEOGRAPHIC</span>
<span>WEIGHTS</span> <span>MATRIX</span> <span>ACCOMPANYING</span>
<span>GEO-REFERENCED</span> <span>DATA</span>.”</span> <em>Can.
Geogr.</em> 40 (4): 351–67.
</div>
<div id="ref-Griffith2000-mn" class="csl-entry">
———. 2000. <span>“A Linear Regression Solution to the Spatial
Autocorrelation Problem.”</span> <em>J. Geogr. Syst.</em> 2 (2): 141–56.
</div>
<div id="ref-Griffith2019-bo" class="csl-entry">
———. 2019. <span>“Negative Spatial Autocorrelation: One of the Most
Neglected Concepts in Spatial Statistics.”</span> <em>Stats</em> 2 (3):
388–415.
</div>
<div id="ref-Griffith2022-lk" class="csl-entry">
Griffith, Daniel A, and Yongwan Chun. 2022. <span>“Some Useful Details
about the Moran Coefficient, the Geary Ratio, and the Join Count Indices
of Spatial Autocorrelation.”</span> <em>Journal of Spatial
Econometrics</em> 3 (1): 12.
</div>
<div id="ref-Griffith2006-kx" class="csl-entry">
Griffith, Daniel A, and Robert Haining. 2006. <span>“Beyond Mule Kicks:
The Poisson Distribution in Geographical Analysis.”</span> <em>Geogr.
Anal.</em> 38 (2): 123–39.
</div>
<div id="ref-Griffith2006-yl" class="csl-entry">
Griffith, Daniel A, and Pedro R Peres-Neto. 2006. <span>“Spatial
Modeling in Ecology: The Flexibility of Eigenfunction Spatial
Analyses.”</span> <em>Ecology</em> 87 (10): 2603–13.
</div>
<div id="ref-De_Jong1984-br" class="csl-entry">
Jong, Peter de, C Sprenger, and Frans van Veen. 1984. <span>“On Extreme
Values of Moran’s <span>I</span> and Geary’s c.”</span> <em>Geogr.
Anal.</em> 16 (1): 17–24.
</div>
<div id="ref-Langfelder2008-fs" class="csl-entry">
Langfelder, Peter, and Steve Horvath. 2008. <span>“<span>WGCNA</span>:
An <span>R</span> Package for Weighted Correlation Network
Analysis.”</span> <em>BMC Bioinformatics</em> 9 (December): 559.
</div>
<div id="ref-Lee2001-tm" class="csl-entry">
Lee, Sang-Il. 2001. <span>“Developing a Bivariate Spatial Association
Measure: An Integration of Pearson’s r and Moran’s
<span>I</span>.”</span> <em>J. Geogr. Syst.</em> 3 (4): 369–85.
</div>
<div id="ref-Lun2016-yq" class="csl-entry">
Lun, Aaron T L, Karsten Bach, and John C Marioni. 2016. <span>“Pooling
Across Cells to Normalize Single-Cell <span>RNA</span> Sequencing Data
with Many Zero Counts.”</span> <em>Genome Biol.</em> 17 (April): 75.
</div>
<div id="ref-Luo2019-dk" class="csl-entry">
Luo, Qing, Daniel A Griffith, and Huayi Wu. 2019. <span>“Spatial
Autocorrelation for Massive Spatial Data: Verification of Efficiency and
Statistical Power Asymptotics.”</span> <em>J. Geogr. Syst.</em> 21 (2):
237–69.
</div>
<div id="ref-McKellar2021-ek" class="csl-entry">
McKellar, David W, Lauren D Walter, Leo T Song, Madhav Mantri, Michael F
Z Wang, Iwijn De Vlaminck, and Benjamin D Cosgrove. 2021.
<span>“Large-Scale Integration of Single-Cell Transcriptomic Data
Captures Transitional Progenitor States in Mouse Skeletal Muscle
Regeneration.”</span> <em>Commun Biol</em> 4 (1): 1280.
</div>
<div id="ref-Miller2021-zs" class="csl-entry">
Miller, Brendan F, Dhananjay Bambah-Mukku, Catherine Dulac, Xiaowei
Zhuang, and Jean Fan. 2021. <span>“Characterizing Spatial Gene
Expression Heterogeneity in Spatially Resolved Single-Cell
Transcriptomic Data with Nonuniform Cellular Densities.”</span>
<em>Genome Res.</em> 31 (10): 1843–55.
</div>
<div id="ref-Moran1950-aq" class="csl-entry">
Moran, P A P. 1950. <span>“Notes on Continuous Stochastic
Phenomena.”</span> <em>Biometrika</em> 37 (1-2): 17–23.
</div>
<div id="ref-Ord2012-qi" class="csl-entry">
Ord, J Keith, and Arthur Getis. 2012. <span>“Local Spatial
Heteroscedasticity (<span>LOSH</span>).”</span> <em>Ann. Reg. Sci.</em>
48 (2): 529–39.
</div>
<div id="ref-Ord1995-we" class="csl-entry">
Ord, J K, and Arthur Getis. 1995. <span>“Local Spatial Autocorrelation
Statistics: Distributional Issues and an Application.”</span> <em>Geogr.
Anal.</em> 27 (4): 286–306.
</div>
<div id="ref-Shang2022-qy" class="csl-entry">
Shang, Lulu, and Xiang Zhou. 2022. <span>“Spatially Aware Dimension
Reduction for Spatial Transcriptomics.”</span> <em>Nat. Commun.</em> 13
(1): 7203.
</div>
<div id="ref-Velten2022-gv" class="csl-entry">
Velten, Britta, Jana M Braunger, Ricard Argelaguet, Damien Arnol, Jakob
Wirbel, Danila Bredikhin, Georg Zeller, and Oliver Stegle. 2022.
<span>“Identifying Temporal and Spatial Patterns of Variation from
Multimodal Data Using <span>MEFISTO</span>.”</span> <em>Nat.
Methods</em> 19 (2): 179–86.
</div>
<div id="ref-Wang2017-li" class="csl-entry">
Wang, Chao, Feng Yue, and Shihuan Kuang. 2017. <span>“Muscle Histology
Characterization Using <span>H&amp;E</span> Staining and Muscle Fiber
Type Classification Using Immunofluorescence Staining.”</span> <em>Bio
Protoc</em> 7 (10).
</div>
<div id="ref-Wartenberg1985-fk" class="csl-entry">
Wartenberg, Daniel. 1985. <span>“Multivariate Spatial Correlation: A
Method for Exploratory Geographical Analysis.”</span> <em>Geogr.
Anal.</em> 17 (4): 263–83.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Alik Huseynov, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
